// @contentstack/management/dist/es-modules/stack/taxonomy/terms/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _regeneratorRuntime from "@babel/runtime/regenerator";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import cloneDeep from 'lodash/cloneDeep';
import { create, fetch, update, query, deleteEntity, move, parseData } from '../../../entity';
import error from '../../../core/contentstackError';

/**
 * Terms are individual items within a taxonomy that allow you to organize and categorize content.
 * @namespace Terms
 */
export function Terms(http, data) {
  var _this = this;
  this.stackHeaders = data.stackHeaders;
  this.taxonomy_uid = data.taxonomy_uid;
  this.urlPath = "/taxonomies/".concat(this.taxonomy_uid, "/terms");
  if (data && data.term) {
    Object.assign(this, cloneDeep(data.term));
    this.urlPath = "/taxonomies/".concat(this.taxonomy_uid, "/terms/").concat(this.uid);

    /**
     * @description The Update term call is used to update an existing term.
     * @memberof Terms
     * @func update
     * @returns {Promise<Terms.Terms>} Promise for Terms instance
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).terms('terms_uid').fetch()
     * .then((terms) => {
     *    terms.name = 'terms name'
     *    return terms.update()
     * })
     * .then((terms) => console.log(terms))
     *
     */
    this.update = update(http, 'term');

    /**
     * @description The Delete term call is used to delete an existing term.
     * @memberof Terms
     * @func delete
     * @returns {Object} Response Object.
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).terms('terms_uid').delete()
     * .then((response) => console.log(response.notice))
     *
     */
    this["delete"] = deleteEntity(http);

    /**
     * @description The Fetch term call is used to fetch an existing term.
     * @memberof Terms
     * @func fetch
     * @returns {Promise<Terms.Terms>} Promise for Terms instance
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).terms('terms_uid').fetch()
     * .then((terms) => console.log(terms))
     *
     */
    this.fetch = fetch(http, 'term');

    /**
     * @description The ancestors call is used to get all the ancestor terms of an existing term.
     * @memberof Terms
     * @func ancestors
     * @returns {Promise<Object>} Promise for ancestors data
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).terms('terms_uid').ancestors()
     * .then((terms) => console.log(terms))
     *
     */
    this.ancestors = /*#__PURE__*/_asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
      var params,
        headers,
        response,
        _args = arguments,
        _t;
      return _regeneratorRuntime.wrap(function (_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            params = _args.length > 0 && _args[0] !== undefined ? _args[0] : {};
            _context.prev = 1;
            headers = {
              headers: _objectSpread(_objectSpread({}, cloneDeep(_this.stackHeaders)), cloneDeep(params))
            };
            _context.next = 2;
            return http.get("".concat(_this.urlPath, "/ancestors"), headers);
          case 2:
            response = _context.sent;
            return _context.abrupt("return", parseData(response, _this.stackHeaders));
          case 3:
            _context.prev = 3;
            _t = _context["catch"](1);
            console.error(_t);
            throw _t;
          case 4:
          case "end":
            return _context.stop();
        }
      }, _callee, null, [[1, 3]]);
    }));

    /**
     * @description The descendants call gets all descendant terms of an existing term.
     * @memberof Terms
     * @func descendants
     * @returns {Promise<Object>} Promise for descendants data
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).terms('terms_uid').descendants()
     * .then((terms) => console.log(terms))
     *
     */
    this.descendants = /*#__PURE__*/_asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee2() {
      var params,
        headers,
        response,
        _args2 = arguments,
        _t2;
      return _regeneratorRuntime.wrap(function (_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            params = _args2.length > 0 && _args2[0] !== undefined ? _args2[0] : {};
            _context2.prev = 1;
            headers = {
              headers: _objectSpread(_objectSpread({}, cloneDeep(_this.stackHeaders)), cloneDeep(params))
            };
            _context2.next = 2;
            return http.get("".concat(_this.urlPath, "/descendants"), headers);
          case 2:
            response = _context2.sent;
            return _context2.abrupt("return", parseData(response, _this.stackHeaders));
          case 3:
            _context2.prev = 3;
            _t2 = _context2["catch"](1);
            console.error(_t2);
            throw _t2;
          case 4:
          case "end":
            return _context2.stop();
        }
      }, _callee2, null, [[1, 3]]);
    }));

    /**
     * @description The move call is used to update the parent uid.
     * @memberof Terms
     * @func move
     * @returns {Promise<Terms.Terms>} Promise for Terms instance
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * const term = {
     *  parent_uid: 'parent_uid',
     *  order: 2
     * }
     * client.stack({ api_key: 'api_key'}).terms('terms_uid').move(term)
     * .then((terms) => console.log(terms))
     *
     */
    this.move = move(http, 'term');

    /**
     * @description The Get term locales call is used to fetch a term in all locales where it's localized.
     * @memberof Terms
     * @func locales
     * @returns {Promise<Object>} Promise for term locales response
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).taxonomy('taxonomy_uid').terms('term_uid').locales()
     * .then((response) => console.log(response.terms))
     *
     */
    this.locales = /*#__PURE__*/_asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee3() {
      var params,
        headers,
        response,
        _args3 = arguments,
        _t3;
      return _regeneratorRuntime.wrap(function (_context3) {
        while (1) switch (_context3.prev = _context3.next) {
          case 0:
            params = _args3.length > 0 && _args3[0] !== undefined ? _args3[0] : {};
            _context3.prev = 1;
            headers = {
              headers: _objectSpread({}, cloneDeep(_this.stackHeaders)),
              params: params
            };
            _context3.next = 2;
            return http.get("".concat(_this.urlPath, "/locales"), headers);
          case 2:
            response = _context3.sent;
            if (!response.data) {
              _context3.next = 3;
              break;
            }
            return _context3.abrupt("return", response.data);
          case 3:
            throw error(response);
          case 4:
            _context3.next = 6;
            break;
          case 5:
            _context3.prev = 5;
            _t3 = _context3["catch"](1);
            throw error(_t3);
          case 6:
          case "end":
            return _context3.stop();
        }
      }, _callee3, null, [[1, 5]]);
    }));

    /**
     * @description The Localize term call is used to create a localized version of a term.
     * @memberof Terms
     * @func localize
     * @param {Object} data - The localization data containing term object
     * @param {Object} params - Query parameters including locale
     * @returns {Promise<Terms.Terms>} Promise for Terms instance
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).taxonomy('taxonomy_UID').terms('term_uid').localize({ term: data }, { locale: 'hi-in' })
     * .then((term) => console.log(term))
     *
     */
    this.localize = /*#__PURE__*/function () {
      var _ref4 = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee4(data) {
        var params,
          headers,
          response,
          _args4 = arguments,
          _t4;
        return _regeneratorRuntime.wrap(function (_context4) {
          while (1) switch (_context4.prev = _context4.next) {
            case 0:
              params = _args4.length > 1 && _args4[1] !== undefined ? _args4[1] : {};
              _context4.prev = 1;
              headers = {
                headers: _objectSpread({}, cloneDeep(_this.stackHeaders))
              };
              _context4.next = 2;
              return http.post(_this.urlPath, data, _objectSpread(_objectSpread({}, headers), {}, {
                params: params
              }));
            case 2:
              response = _context4.sent;
              if (!response.data) {
                _context4.next = 3;
                break;
              }
              return _context4.abrupt("return", new _this.constructor(http, parseData(response, _this.stackHeaders)));
            case 3:
              throw error(response);
            case 4:
              _context4.next = 6;
              break;
            case 5:
              _context4.prev = 5;
              _t4 = _context4["catch"](1);
              throw error(_t4);
            case 6:
            case "end":
              return _context4.stop();
          }
        }, _callee4, null, [[1, 5]]);
      }));
      return function (_x) {
        return _ref4.apply(this, arguments);
      };
    }();
  } else {
    /**
     * @description The Create term call creates a new term in a taxonomy.
     * @memberof Terms
     * @func create
     * @returns {Promise<Terms.Terms>} Promise for Terms instance
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     * const terms = {
     *    uid: 'terms_testing1',
     *    name: 'terms testing',
     *    description: 'Description for terms testing'
     * }
     * client.stack({ api_key: 'api_key'}).terms().create({ terms })
     * .then((terms) => console.log(terms))
     *
     */
    this.create = create({
      http: http
    });

    /**
     * @description The Query on Terms will allow you to fetch details of all Terms.
     * @memberof Terms
     * @param {Object} params - URI parameters
     * @prop {Object} params.query - Queries that you can use to fetch filtered results.
     * @func query
     * @returns {Object} Query builder object with find(), count(), and findOne() methods.
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack().terms().query().find()
     * .then((terms) => console.log(terms))
     */
    this.query = query({
      http: http,
      wrapperCollection: TermsCollection
    });
  }
  /**
     * @description The Search terms call is used to search a term.
     * @memberof Terms
     * @func search
     * @returns {Promise<Object>} Promise for search results
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     * const term_string = ''
     * client.stack({ api_key: 'api_key'}).taxonomy('taxonomy_uid').terms().search(term_string)
     * .then((terms) => console.log(terms))
     *
     */
  this.search = /*#__PURE__*/_asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee5() {
    var term,
      params,
      headers,
      response,
      _args5 = arguments,
      _t5;
    return _regeneratorRuntime.wrap(function (_context5) {
      while (1) switch (_context5.prev = _context5.next) {
        case 0:
          term = _args5.length > 0 && _args5[0] !== undefined ? _args5[0] : '';
          params = _args5.length > 1 && _args5[1] !== undefined ? _args5[1] : {};
          _context5.prev = 1;
          headers = {
            headers: _objectSpread(_objectSpread({}, cloneDeep(_this.stackHeaders)), cloneDeep(params))
          };
          _context5.next = 2;
          return http.get("taxonomies/$all/terms?typeahead=".concat(term), headers);
        case 2:
          response = _context5.sent;
          return _context5.abrupt("return", parseData(response, _this.stackHeaders));
        case 3:
          _context5.prev = 3;
          _t5 = _context5["catch"](1);
          console.error(_t5);
          throw _t5;
        case 4:
        case "end":
          return _context5.stop();
      }
    }, _callee5, null, [[1, 3]]);
  }));
}
export function TermsCollection(http, data) {
  var obj = cloneDeep(data.terms) || [];
  var termsCollection = obj.map(function (term) {
    return new Terms(http, {
      term: term,
      taxonomy_uid: data.taxonomy_uid,
      stackHeaders: data.stackHeaders
    });
  });
  return termsCollection;
}
