// @contentstack/management/dist/es-modules/stack/variantGroup/variants/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _regeneratorRuntime from "@babel/runtime/regenerator";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import cloneDeep from 'lodash/cloneDeep';
import { query, deleteEntity } from '../../../entity';
import error from '../../../core/contentstackError';

/**
 * Variants within a variant group allow you to manage individual variants that belong to a specific variant group. Read more about <a href='https://www.contentstack.com/docs/developers/create-content-types/manage-variants'>Variants</a>.
 * @namespace Variants
 */

export function Variants(http) {
  var _this = this;
  var data = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  Object.assign(this, cloneDeep(data.variants));
  this.stackHeaders = data.stackHeaders;
  this.variant_group_uid = data.variant_group_uid;
  this.urlPath = "/variant_groups/".concat(this.variant_group_uid, "/variants");
  if (data.variants) {
    this.urlPath += "/".concat(this.uid);
    /**
       * @description The Update Variant call lets you update the name and description of an existing Variant.
       * @memberof Variants
       * @func update
       * @returns {Promise<Object>} Response Object.
       * @example
       * import * as contentstack from '@contentstack/management'
       * const client = contentstack.client()
       * const data = { "name": "Update name" }
       * client.stack({ api_key: 'api_key'}).variantGroup('variant_group_uid').variants('variant_uid').update(data)
       * .then((variants) => console.log(variants))
       *
       */
    this.update = /*#__PURE__*/function () {
      var _ref = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee(data) {
        var response, _t;
        return _regeneratorRuntime.wrap(function (_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _context.prev = 0;
              _context.next = 1;
              return http.put(_this.urlPath, data, {
                headers: _objectSpread({}, cloneDeep(_this.stackHeaders))
              });
            case 1:
              response = _context.sent;
              if (!response.data) {
                _context.next = 2;
                break;
              }
              return _context.abrupt("return", response.data);
            case 2:
              return _context.abrupt("return", error(response));
            case 3:
              _context.next = 5;
              break;
            case 4:
              _context.prev = 4;
              _t = _context["catch"](0);
              return _context.abrupt("return", error(_t));
            case 5:
            case "end":
              return _context.stop();
          }
        }, _callee, null, [[0, 4]]);
      }));
      return function (_x) {
        return _ref.apply(this, arguments);
      };
    }();

    /**
       * @description The fetch Variant call fetches Variant details.
       * @memberof Variants
       * @func fetch
       * @returns {Promise<Object>} Response Object.
       * @param {Object} param - Query parameters for fetching variant details
       * @example
       * import * as contentstack from '@contentstack/management'
       * const client = contentstack.client()
       *
       * client.stack({ api_key: 'api_key'}).variantGroup('variant_group_uid').variants('variant_uid').fetch()
       * .then((variants) => console.log(variants))
       *
       */
    this.fetch = /*#__PURE__*/_asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee2() {
      var param,
        headers,
        response,
        _args2 = arguments,
        _t2;
      return _regeneratorRuntime.wrap(function (_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            param = _args2.length > 0 && _args2[0] !== undefined ? _args2[0] : {};
            _context2.prev = 1;
            headers = {
              headers: _objectSpread({}, cloneDeep(_this.stackHeaders)),
              params: _objectSpread({}, cloneDeep(param))
            } || {};
            _context2.next = 2;
            return http.get(_this.urlPath, headers);
          case 2:
            response = _context2.sent;
            if (!response.data) {
              _context2.next = 3;
              break;
            }
            return _context2.abrupt("return", response.data);
          case 3:
            throw error(response);
          case 4:
            _context2.next = 6;
            break;
          case 5:
            _context2.prev = 5;
            _t2 = _context2["catch"](1);
            error(_t2);
          case 6:
          case "end":
            return _context2.stop();
        }
      }, _callee2, null, [[1, 5]]);
    }));

    /**
       * @description The Delete Variant call is used to delete an existing Variant permanently from your Variant Group.
       * @memberof Variants
       * @func delete
       * @returns {Promise<Object>} Response Object.
       * @example
       * import * as contentstack from '@contentstack/management'
       * const client = contentstack.client()
       *
       * client.stack({ api_key: 'api_key'}).variantGroup('variant_group_uid').variants('variant_uid').delete()
       * .then((response) => console.log(response.notice))
       */
    this["delete"] = deleteEntity(http);
  } else {
    /**
     * @description The Create a variant call creates a new variant in a particular Variant Group.
     * @memberof Variants
     * @func create
     * @returns {Promise<Object>} Response Object.
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     * const data = {
     *      "uid": "iphone_color_white", // optional
     *      "name": "White",
     *      "personalize_metadata": {   // optional sent from personalize while creating variant
     *          "experience_uid": "exp1",
     *          "experience_short_uid": "expShortUid1",
     *          "project_uid": "project_uid1",
     *          "variant_short_uid": "variantShort_uid1"
     *      },
     *      }
     * client.stack().variantGroup('variant_group_uid').variants().create({ data })
     * .then((variants) => console.log(variants))
     */
    this.create = /*#__PURE__*/function () {
      var _ref3 = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee3(data) {
        var response, _t3;
        return _regeneratorRuntime.wrap(function (_context3) {
          while (1) switch (_context3.prev = _context3.next) {
            case 0:
              _context3.prev = 0;
              _context3.next = 1;
              return http.post("".concat(_this.urlPath), data, {
                headers: _objectSpread({}, cloneDeep(_this.stackHeaders))
              });
            case 1:
              response = _context3.sent;
              if (!response.data) {
                _context3.next = 2;
                break;
              }
              return _context3.abrupt("return", response.data);
            case 2:
              return _context3.abrupt("return", error(response));
            case 3:
              _context3.next = 5;
              break;
            case 4:
              _context3.prev = 4;
              _t3 = _context3["catch"](0);
              return _context3.abrupt("return", error(_t3));
            case 5:
            case "end":
              return _context3.stop();
          }
        }, _callee3, null, [[0, 4]]);
      }));
      return function (_x2) {
        return _ref3.apply(this, arguments);
      };
    }();

    /**
     * @description The Query on Variants will allow you to fetch details of all or specific Variants.
     * @memberof Variants
     * @func query
     * @param {Object} params - URI parameters
     * @prop {Object} params.query - Queries that you can use to fetch filtered results for variants.
     * @returns {Object} Query builder object with find(), count(), and findOne() methods.
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
      * client.stack({ api_key: 'api_key' }).variantGroup('variant_group_uid').variants().query({ query: { name: 'white' } }).find()
     * .then((variants) => console.log(variants))
     */
    this.query = query({
      http: http,
      wrapperCollection: VariantsCollection
    });
  }
  return this;
}
export function VariantsCollection(http, data) {
  var obj = cloneDeep(data.variants) || [];
  var variantsCollection = obj.map(function (userdata) {
    return new Variants(http, {
      variants: userdata,
      variant_group_uid: data.uid,
      stackHeaders: data.stackHeaders
    });
  });
  return variantsCollection;
}
