// @contentstack/management/dist/es5/user/index.js
"use strict";

var _interopRequireDefault3 = require("@babel/runtime/helpers/interopRequireDefault");
var _interopRequireDefault2 = _interopRequireDefault3(require("@babel/runtime/helpers/interopRequireDefault"));
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _defineProperty2 = require("@babel/runtime/helpers/defineProperty");
var _defineProperty3 = (0, _interopRequireDefault2["default"])(_defineProperty2);
var _asyncToGenerator2 = require("@babel/runtime/helpers/asyncToGenerator");
var _asyncToGenerator3 = (0, _interopRequireDefault2["default"])(_asyncToGenerator2);
exports.User = User;
exports.UserCollection = UserCollection;
var _regenerator = require("@babel/runtime/regenerator");
var _regenerator2 = (0, _interopRequireDefault2["default"])(_regenerator);
var _cloneDeep = require("lodash/cloneDeep");
var _cloneDeep2 = (0, _interopRequireDefault2["default"])(_cloneDeep);
var _entity = require("../entity");
var _contentstackError = require("../core/contentstackError");
var _contentstackError2 = (0, _interopRequireDefault2["default"])(_contentstackError);
var _organization = require("../organization");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty3["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
/**
 * All accounts registered with Contentstack are known as Users. A stack can have many users with varying permissions and roles. Read Users to learn more.
 * @namespace User
 */
function User(http, data) {
  Object.assign(this, (0, _cloneDeep2["default"])(data.user));
  this.urlPath = '/user';
  if (this.authtoken) {
    /**
     * @description The Update User API Request updates the details of an existing user account.
     * @memberof User
     * @func update
     * @returns {Promise<User.User>} Promise for User instance
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).getUser()
     * .then((user) => {
     *  user.first_name = 'FirstName'
     *  user.last_name = 'LastName'
     *  user.company = 'Company Name'
     *  return user.update()
     * })
     * .then((user) => console.log(user))
     */
    this.update = (0, _entity.update)(http, 'user');

    /**
     * @description The Delete user call deletes the current authenticated user permanently from your Contentstack account.
     * @memberof User
     * @func delete
     * @returns {Object} Response Object.
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).getUser()
     * .then((user) => {
     *  return user.delete()
     * })
     * .then((response) => console.log(response.notice))
     */
    this["delete"] = (0, _entity.deleteEntity)(http);

    /**
     * @description The Request for a password call sends a request for a temporary password to log in to an account in case a user has forgotten the login password.
     * @memberof User
     * @func requestPassword
     * @returns {Object} Response Object.
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).getUser()
     * .then((user) => {
     *  return user.requestPassword()
     * })
     * .then((response) => console.log(response.notice))
     *
     */
    this.requestPassword = function () {
      return http.post('/user/forgot_password', {
        user: {
          email: this.email
        }
      }).then(function (response) {
        return response.data;
      }, _contentstackError2["default"]);
    };

    /**
     * @description The Reset password call sends a request for resetting the password of your Contentstack account.
     * @memberof User
     * @func resetPassword
     * @param {Object} credentials - Password reset credentials.
     * @prop {string} credentials.resetPasswordToken - Reset password token.
     * @prop {string} credentials.password - New password.
     * @prop {string} credentials.passwordConfirm - Password confirmation.
     * @returns {Object} Response Object.
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).getUser()
     * .then((user) => {
     *  return user.resetPassword({ resetPasswordToken: 'token', password: 'new_password', passwordConfirm: 'new_password' })
     * })
     * .then((response) => console.log(response.notice))
     *
     */
    this.resetPassword = function (_ref) {
      var resetPasswordToken = _ref.resetPasswordToken,
        password = _ref.password,
        passwordConfirm = _ref.passwordConfirm;
      return http.post('/user/reset_password', {
        user: {
          reset_password_token: resetPasswordToken,
          password: password,
          password_confirmation: passwordConfirm
        }
      }).then(function (response) {
        return response.data;
      }, _contentstackError2["default"]);
    };
    if (this.organizations) {
      this.organizations = new _organization.OrganizationCollection(http, {
        organizations: this.organizations
      });
    }

    /**
     * @description The Get all Tasks request retrieves a list of all tasks assigned to you.
     * @memberof User
     * @func getTasks
     * @async
     * @param {Object} params - Query parameters
     * @prop {Object} params.query - Enter the actual query that will be executed to retrieve the tasks. This query should be in JSON format.
     * @prop {Object} params.sort - Enter the field UID on the basis of which you want to sort your tasks.
     * @prop {Int} params.limit - Enter the maximum number of tasks that you want to retrieve in the response.
     * @prop {Int} params.skip - Enter the number of tasks to be skipped.
     * @returns {Promise<Object>} Response Object.
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).getUser()
     * .then((user) => {
     *  return user.getTasks()
     * })
     * .then((response) => console.log(response.assignments))
     *
     */

    this.getTasks = /*#__PURE__*/function () {
      var _ref2 = (0, _asyncToGenerator3["default"])(/*#__PURE__*/_regenerator2["default"].mark(function _callee(params) {
        var headers, response, _t;
        return _regenerator2["default"].wrap(function (_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              headers = {};
              if (params) {
                headers.params = _objectSpread({}, (0, _cloneDeep2["default"])(params));
              }
              _context.prev = 1;
              _context.next = 2;
              return http.get("/user/assignments", headers);
            case 2:
              response = _context.sent;
              if (!response.data) {
                _context.next = 3;
                break;
              }
              return _context.abrupt("return", response.data);
            case 3:
              throw (0, _contentstackError2["default"])(response);
            case 4:
              _context.next = 6;
              break;
            case 5:
              _context.prev = 5;
              _t = _context["catch"](1);
              throw (0, _contentstackError2["default"])(_t);
            case 6:
            case "end":
              return _context.stop();
          }
        }, _callee, null, [[1, 5]]);
      }));
      return function (_x) {
        return _ref2.apply(this, arguments);
      };
    }();
  }
  return this;
}
function UserCollection(http, data) {
  var users = data.collaborators || data.shares || [];
  var obj = (0, _cloneDeep2["default"])(users);
  var userCollection = obj.map(function (userdata) {
    return new User(http, {
      user: userdata
    });
  });
  return userCollection;
}
