// @contentstack/venus-components/build/variables.js
/**
 * Do not edit directly
 * Generated on Fri, 14 Jul 2023 15:03:53 GMT
 */

export const ShadowDefault = "0 1px 3px rgba(0, 0, 0, 0.08)";
export const ShadowHeavy = "0 2px 3px rgba(0, 0, 0, 0.35)";
export const ShadowThemeLightAmbiencePurpleShadow1Px = "0px 0px 2px rgba(0, 0, 0, 0.14), 0px 2px 2px rgba(0, 0, 0, 0.12), 0px 1px 3px rgba(108, 92, 231, 0.2)";
export const ShadowThemeLightAmbiencePurpleShadow4Px = "0px 2px 4px rgba(0, 0, 0, 0.14), 0px 5px 5px rgba(0, 0, 0, 0.12), 0px 1px 10px rgba(108, 92, 231, 0.2)";
export const ShadowThemeLightAmbiencePurpleShadow8Px = "0px 8px 10px 1px rgba(0, 0, 0, 0.14), 0px 3px 14px 3px rgba(0, 0, 0, 0.12), 0px 4px 15px rgba(108, 92, 231, 0.2)";
export const ShadowThemeLightAmbiencePurpleShadow16Px = "0px 16px 24px 2px rgba(0, 0, 0, 0.14), 0px 6px 30px 5px rgba(0, 0, 0, 0.12), 0px 8px 10px 3px rgba(108, 92, 231, 0.2)";
export const ShadowThemeLightAmbiencePurpleShadow24Px = "0px 24px 38px 3px rgba(0, 0, 0, 0.14), 0px 9px 46px 8px rgba(0, 0, 0, 0.12), 0px 11px 15px 8px rgba(108, 92, 231, 0.2)";
export const ShadowThemeLightAmbienceBlackShadow1Px = "0px 0px 2px rgba(0, 0, 0, 0.14), 0px 2px 2px rgba(0, 0, 0, 0.12), 0px 1px 3px rgba(0, 0, 0, 0.2)";
export const ShadowThemeLightAmbienceBlackShadow4Px = "0px 2px 4px rgba(0, 0, 0, 0.14), 0px 5px 5px rgba(0, 0, 0, 0.12), 0px 1px 10px rgba(0, 0, 0, 0.2)";
export const ShadowThemeLightAmbienceBlackShadow8Px = "0px 8px 10px 1px rgba(0, 0, 0, 0.14), 0px 3px 14px 3px rgba(0, 0, 0, 0.12), 0px 4px 15px rgba(0, 0, 0, 0.2)";
export const ShadowThemeLightAmbienceBlackShadow16Px = "0px 16px 24px 2px rgba(0, 0, 0, 0.14), 0px 6px 30px 5px rgba(0, 0, 0, 0.12), 0px 8px 10px rgba(0, 0, 0, 0.2)";
export const ShadowThemeLightAmbienceBlackShadow24Px = "0px 24px 38px 3px rgba(0, 0, 0, 0.14), 0px 9px 46px 8px rgba(0, 0, 0, 0.12), 0px 11px 15px rgba(0, 0, 0, 0.2)";
export const ColorBasePurple5 = "#6c5ce7";
export const ColorBasePurple10 = "#3e3871";
export const ColorBasePurple15 = "#efedfc";
export const ColorBasePurple20 = "#9387ed";
export const ColorBasePurple25 = "#5d50bf";
export const ColorBaseGray5 = "#edf1f7";
export const ColorBaseGray10 = "#a9b6cb";
export const ColorBaseGray15 = "#647696";
export const ColorBaseGray20 = "#475161";
export const ColorBaseGray25 = "#a0aec0";
export const ColorBaseGray30 = "#f5f5f5";
export const ColorBaseGray35 = "#c7d0e1";
export const ColorBaseGray40 = "#dde3ee";
export const ColorBaseGreen5 = "#007a52";
export const ColorBaseGreen10 = "#00472f";
export const ColorBaseGreen15 = "#f5fffc";
export const ColorBaseOrange5 = "#d62400";
export const ColorBaseOrange10 = "#701300";
export const ColorBaseOrange15 = "#ffeeeb";
export const ColorBaseOrange20 = "#eb5646";
export const ColorBaseBlue5 = "#00b9e0";
export const ColorBaseBlue10 = "#003b47";
export const ColorBaseBlue15 = "#f5fdff";
export const ColorBaseYellow5 = "#ffae0a";
export const ColorBaseYellow10 = "#704b00";
export const ColorBaseYellow15 = "#fff8eb";
export const ColorBaseBlack5 = "#222222";
export const ColorBaseWhite5 = "#ffffff";
export const ColorBaseWhite10 = "#f7f9fc";
export const ColorBaseWhite15 = "#f5f5f5";
export const ColorBaseWhite20 = "#f7f9fc";
export const ColorStatusAccentsPoppyR600 = "#a31b00";
export const ColorStatusAccentsMangoY500 = "#ffae0a";
export const ColorStatusAccentsSpinachG800 = "#007a52";
export const ColorStatusAccentsAquaB600 = "#0469e3";
export const ColorMaximumBluePurple = "#ada4f4";
export const ColorAccentColorsPurple50 = "#b6aef3";
export const ColorAccentColorsMediumOrchid = "#bd59fa";
export const ColorAccentColorsMaximumBluePurple = "#000000";
export const ColorAccentColorsBlueChalk = "#edebff";
export const ColorAccentColorsPaleAqua = "#bad4fb";
export const ColorAccentColorsPastelBlue = "#43b7c2";
export const ColorAccentColorsBottleGreen = "#003a25";
export const ColorAccentColorsAquaDeep = "#00593b";
export const ColorAccentColorsMint = "#47b0aa";
export const ColorAccentColorsSummerGreen = "#90bba5";
export const ColorAccentColorsBean = "#4f0800";
export const ColorAccentColorsRustOrange = "#ba5800";
export const ColorAccentColorsSalmonPink = "#e57872";
export const ColorAccentColorsCrayola = "#ffce6c";
export const ColorAccentColorsSundown = "#fdafa3";
export const ColorAccentColorsButterscotch = "#ffad48";
export const ColorBorderLighter = "#f5f5f5";
export const ColorBorderLight = "#a9b6cb";
export const ColorBorderBase = "#647696";
export const ColorBorderDark = "#475161";
export const ColorBorderLink = "#6c5ce7";
export const ColorBorderFocus = "#edf1f7";
export const ColorBorderWarning = "#d62400";
export const ColorBorderSuccess = "#007a52";
export const ColorBrandPrimaryBase = "#6c5ce7";
export const ColorBrandPrimaryMedium = "#9387ed";
export const ColorBrandPrimaryDark = "#3e3871";
export const ColorBrandPrimaryDarker = "#5d50bf";
export const ColorBrandPrimaryLight = "#efedfc";
export const ColorBrandSecondaryLightest = "#dde3ee";
export const ColorBrandSecondaryLighter = "#edf1f7";
export const ColorBrandSecondaryLight = "#a9b6cb";
export const ColorBrandSecondaryBase = "#647696";
export const ColorBrandSecondaryMedium = "#c7d0e1";
export const ColorBrandSecondaryDark = "#475161";
export const ColorBrandSuccessBase = "#007a52";
export const ColorBrandSuccessDark = "#00472f";
export const ColorBrandSuccessLight = "#f5fffc";
export const ColorBrandWarningBase = "#d62400";
export const ColorBrandWarningDark = "#701300";
export const ColorBrandWarningLight = "#ffeeeb";
export const ColorBrandWarningMedium = "#eb5646";
export const ColorBrandAttentionBase = "#ffae0a";
export const ColorBrandAttentionDark = "#704b00";
export const ColorBrandAttentionLight = "#fff8eb";
export const ColorBrandInfoBase = "#00b9e0";
export const ColorBrandInfoDark = "#003b47";
export const ColorBrandInfoLight = "#f5fdff";
export const ColorBrandBlackBase = "#222222";
export const ColorBrandWhiteLight = "#f7f9fc";
export const ColorBrandWhiteBase = "#ffffff";
export const ColorBrandWhiteDark = "#f7f9fc";
export const ColorBrandWhiteDarker = "#f5f5f5";
export const ColorBgColorsLavender = "#f8f6ff";
export const ColorBgColorsDolphinGreyN200 = "#dde3ee";
export const ColorBgColorsPureWhite = "#ffffff";
export const ColorBgColorsMilkybarWhiteN5 = "#f7f9fc";
export const ColorBgColorsHibiscusR5 = "#ffeeeb";
export const ColorBgColorsCashewY5 = "#fff8eb";
export const ColorBgColorsAlgaeG5 = "#f5fffc";
export const ColorBgColorsSkyB5 = "#f5fdff";
export const ColorBgColorsTulipPurpleP5 = "#f9f8ff";
export const ColorBrandColorsBrandOrange = "#eb5646";
export const ColorBrandColorsOrchidPurpleP400Primary = "#6c5ce7";
export const ColorBrandColorsLavenderPurpleP500 = "#5d50be";
export const ColorFontBase = "#475161";
export const ColorFontSecondary = "#647696";
export const ColorFontTertiary = "#a9b6cb";
export const ColorFontQuaternary = "#edf1f7";
export const ColorFontLink = "#6c5ce7";
export const ColorFontActive = "#647696";
export const ColorFontWarning = "#d62400";
export const ColorFontAttention = "#ffae0a";
export const ColorFontSuccess = "#007a52";
export const ColorFontInfo = "#00b9e0";
export const ColorFontDisabled = "#a0aec0";
export const ColorFontWhite = "#ffffff";
export const ColorFontBlack = "#222222";
export const ColorFontTextTertiaryTextSonicSilver = "#767676";
export const ColorFontTextSecondaryTextTarmacGreyN700 = "#475161";
export const ColorFontTextPrimaryTextGray900 = "#212121";
export const ColorFontTextPrimaryTextJetBlack = "#000000";
export const ColorFontTextPrimaryTextPurpleGray = "#6e6b86";
export const LineHeightDefault = "1.5";
export const LineHeightCondensed = "1.25";
export const LineHeightReset = "1";
export const MqLarge = "only screen and (min-width: 64.0625rem)"; // Large form factor: devices larger than tablet
export const MqMedium = "only screen and (min-width: 48rem)"; // Medium form factor: devices larger than phone
export const MqSmall = "only screen and (max-width: 47.9375rem)"; // Small form factor: devices smaller than tablet
export const Opacity0 = "0";
export const Opacity1 = "1";
export const Opacity2 = "0.2";
export const Opacity4 = "0.4";
export const Opacity5 = "0.5";
export const Opacity8 = "0.8";
export const Radii2 = "2px";
export const Radii4 = "4px";
export const Radii6 = "6px";
export const Radii8 = "8px";
export const Radii10 = "10px";
export const Radii50 = "50%";
export const SizeFontTiny = "0.6875rem"; // 11px
export const SizeFontSmall = "0.75rem"; // 12px[Title 2]
export const SizeFontMedium = "0.8125rem"; // 13px
export const SizeFontLarge = "0.875rem"; // 14px[Title 1]
export const SizeFontXl = "1rem"; // 16px[H6]
export const SizeFont2Xl = "1.125rem"; // 18px[H5]
export const SizeFont3Xl = "1.375rem"; // 22px[H4]
export const SizeFont4Xl = "1.875rem"; // 30px[H3]
export const SizeFont5Xl = "2.25rem"; // 36px[H2]
export const SizeFont6Xl = "2.625rem"; // 42px[H1]
export const SizeFontV2Small = "0.75rem"; // 12px[P3]
export const SizeFontV2Medium = "0.875rem"; // 14px[H3|P2]
export const SizeFontV2Large = "1rem"; // 16px[P1]
export const SizeFontV2Xl = "1.25rem"; // 20px[H2]
export const SizeFontV22Xl = "1.75rem"; // 28px[H1]
export const SizeFontV23Xl = "2.125rem"; // 34px[H0]
export const Space2 = "0.125rem"; // 2px
export const Space4 = "0.25rem"; // 4px
export const Space5 = "0.3125rem"; // 5px
export const Space6 = "0.375rem"; // 6px
export const Space8 = "0.5rem"; // 8px
export const Space10 = "0.625rem"; // 10px
export const Space12 = "0.75rem"; // 12px
export const Space14 = "0.875rem"; // 14px
export const Space15 = "0.9375rem"; // 15px
export const Space16 = "1rem"; // 16px
export const Space18 = "1.125rem"; // 18px
export const Space20 = "1.25rem"; // 20px
export const Space22 = "1.375rem"; // 22px
export const Space24 = "1.5rem"; // 24px
export const Space25 = "1.5625rem"; // 25px
export const Space26 = "1.625rem"; // 26px
export const Space28 = "1.75rem"; // 28px
export const Space30 = "1.875rem"; // 30px
export const Space32 = "2rem"; // 32px
export const Space34 = "2.125rem"; // 34px
export const Space35 = "2.1875rem"; // 35px
export const Space36 = "2.25rem"; // 36px
export const Space38 = "2.375rem"; // 38px
export const Space40 = "2.5rem"; // 40px
export const Space50 = "3.125rem"; // 50px
export const Space60 = "3.75rem"; // 60px
export const Space80 = "5rem"; // 80px
export const Space100 = "6.25rem"; // 100px
export const Space160 = "10rem"; // 160px
export const Space001Em = "0.01em"; // 0.01em
export const Space002Em = "0.02em"; // 0.02em
export const FontBase = "16";
export const FontWeightRegular = "400";
export const FontWeightMedium = "500";
export const FontWeightSemiBold = "600";
export const FontWeightBold = "700";
export const FontWeightExtraBold = "800";
export const FontFamilyPrimary = "Inter";
export const ZIndexDefault = "0";
export const ZIndexNegative = "-1";
export const ZIndexDeepdive = "-99999";
