// @contentstack/management/dist/es-modules/stack/branch/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _regeneratorRuntime from "@babel/runtime/regenerator";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import cloneDeep from 'lodash/cloneDeep';
import { create, query, fetch, deleteEntity } from '../../entity';
import { Compare } from './compare';
import { MergeQueue } from './mergeQueue';
import error from '../../core/contentstackError';

/**
 *
 * @namespace Branch
 */
export function Branch(http) {
  var _this = this;
  var data = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  this.stackHeaders = data.stackHeaders;
  this.urlPath = "/stacks/branches";
  data.branch = data.branch || data.branch_alias;
  delete data.branch_alias;
  if (data.branch) {
    Object.assign(this, cloneDeep(data.branch));
    this.urlPath = "/stacks/branches/".concat(this.uid);

    /**
     * @description The Delete Branch call is used to delete an existing Branch permanently from your Stack.
     * @memberof Branch
     * @func delete
     * @returns {Promise<Object>} Response Object.
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).branch('branch_uid').delete()
     * .then((response) => console.log(response.notice))
     */
    this["delete"] = deleteEntity(http, true);

    /**
     * @description The fetch Branch call fetches Branch details.
     * @memberof Branch
     * @func fetch
     * @returns {Promise<Branch>} Promise for Branch instance
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).branch('branch_uid').fetch()
     * .then((branch) => console.log(branch))
     *
     */
    this.fetch = fetch(http, 'branch');

    /**
     * @description Compare allows you to compare any or specific ContentType or GlobalFields.
     * @memberof Branch
     * @func compare
     * @returns {Compare} Instance of Compare.
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).branch('branch_uid').compare('compare_uid')
     *
     */
    this.compare = function (compareBranchUid) {
      var compareData = {
        stackHeaders: _this.stackHeaders
      };
      if (compareBranchUid) {
        compareData.branches = {
          base_branch: _this.uid,
          compare_branch: compareBranchUid
        };
      }
      return new Compare(http, compareData);
    };

    /**
     * @description The Update Branch Settings call updates the settings of an existing Branch.
     * @memberof Branch
     * @func updateSettings
     * @param {IBranchSettings} payload - The settings payload to update.
     * @returns {Promise<Object>} Response Object.
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * const payload = {
     *   branch: {
     *     settings: {
     *       am_v2: {
     *         linked_workspaces: [
     *           { uid: 'main', space_uid: 'space_uid', is_default: true }
     *         ]
     *       }
     *     }
     *   }
     * }
     * client.stack({ api_key: 'api_key' }).branch('branch_uid').updateSettings(payload)
     * .then((response) => console.log(response.notice))
     */
    this.updateSettings = /*#__PURE__*/function () {
      var _ref = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee(payload) {
        var response, _t;
        return _regeneratorRuntime.wrap(function (_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _context.prev = 0;
              _context.next = 1;
              return http.post("".concat(_this.urlPath, "/settings"), payload, {
                headers: _objectSpread({}, cloneDeep(_this.stackHeaders))
              });
            case 1:
              response = _context.sent;
              if (!response.data) {
                _context.next = 2;
                break;
              }
              return _context.abrupt("return", response.data);
            case 2:
              throw error(response);
            case 3:
              _context.next = 5;
              break;
            case 4:
              _context.prev = 4;
              _t = _context["catch"](0);
              throw error(_t);
            case 5:
            case "end":
              return _context.stop();
          }
        }, _callee, null, [[0, 4]]);
      }));
      return function (_x) {
        return _ref.apply(this, arguments);
      };
    }();
  } else {
    /**
     * @description The Create a Branch call creates a new branch in a particular stack of your Contentstack account.
     * @memberof Branch
     * @func create
     * @returns {Promise<Branch>} Promise for Branch instance
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     * const branch =   {
     *      name: 'branch_name',
     *      source: 'master'
     * }
     * client.stack({ api_key: 'api_key'}).branch().create({ branch })
     * .then((branch) => { console.log(branch) })
     */
    this.create = create({
      http: http
    });

    /**
     * @description The 'Get all Branch' request returns comprehensive information about branches created in a Stack.
     * @memberof Branch
     * @func query
     * @returns {Object} Query builder object with find(), count(), and findOne() methods.
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).branch().query().find()
     * .then((collection) => { console.log(collection) })
     */
    this.query = query({
      http: http,
      wrapperCollection: BranchCollection
    });

    /**
     * @description Merge allows user to merge branches in a Stack.
     * @memberof Branch
     * @func merge
     * @async
     * @returns {Promise<Object>} Response Object.
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * const params = {
     *   base_branch: "main",
     *   compare_branch: "dev",
     *   default_merge_strategy: "merge_prefer_base",
     *   merge_comment: "Merging dev into main",
     *   no_revert: true,
     * }
     * const mergeObj = {
     *   item_merge_strategies: [
     *     {
     *       uid: "global_field_uid",
     *       type: "global_field",
     *       merge_strategy: "merge_prefer_base"
     *     }
     *   ],
     * }
     *
     * client.stack({ api_key: 'api_key'}).branch().merge(mergeObj, params)
     */
    this.merge = /*#__PURE__*/function () {
      var _ref2 = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee2(mergeObj, params) {
        var url, header, response, _t2;
        return _regeneratorRuntime.wrap(function (_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              url = '/stacks/branches_merge';
              header = {
                headers: _objectSpread({}, cloneDeep(_this.stackHeaders)),
                params: params
              };
              _context2.prev = 1;
              _context2.next = 2;
              return http.post(url, mergeObj, header);
            case 2:
              response = _context2.sent;
              if (!response.data) {
                _context2.next = 3;
                break;
              }
              return _context2.abrupt("return", response.data);
            case 3:
              throw error(response);
            case 4:
              _context2.next = 6;
              break;
            case 5:
              _context2.prev = 5;
              _t2 = _context2["catch"](1);
              throw error(_t2);
            case 6:
            case "end":
              return _context2.stop();
          }
        }, _callee2, null, [[1, 5]]);
      }));
      return function (_x2, _x3) {
        return _ref2.apply(this, arguments);
      };
    }();

    /**
     * @description Merge Queue provides list of all recent merge jobs in a Stack.
     * @memberof Branch
     * @func mergeQueue
     * @returns {MergeQueue} Instance of MergeQueue
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).branch().mergeQueue()
     *
     */
    this.mergeQueue = function (uid) {
      var mergeData = {
        stackHeaders: _this.stackHeaders,
        uid: uid
      };
      return new MergeQueue(http, mergeData);
    };
  }
  return this;
}
export function BranchCollection(http, data) {
  var obj = cloneDeep(data.branches) || data.branch_aliases || [];
  return obj.map(function (branchData) {
    return new Branch(http, {
      branch: branchData,
      stackHeaders: data.stackHeaders
    });
  });
}
