// @contentstack/management/dist/es5/core/pkceStorage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getStoredCodeVerifier = getStoredCodeVerifier;
exports.storeCodeVerifier = storeCodeVerifier;
exports.clearStoredCodeVerifier = clearStoredCodeVerifier;
/**
 * PKCE code_verifier persistence in sessionStorage for browser SPAs.
 * Survives OAuth redirects; not used in Node. RFC 7636 / OAuth 2.0 for Browser-Based Apps.
 */

var PKCE_STORAGE_KEY_PREFIX = 'contentstack_oauth_pkce';
var PKCE_STORAGE_EXPIRY_MS = 10 * 60 * 1000; // 10 minutes

function isBrowser() {
  return typeof window !== 'undefined' && typeof window.sessionStorage !== 'undefined';
}
function getStorageKey(appId, clientId, redirectUri) {
  return "".concat(PKCE_STORAGE_KEY_PREFIX, "_").concat(appId, "_").concat(clientId, "_").concat(redirectUri);
}

/**
 * @param {string} appId
 * @param {string} clientId
 * @param {string} redirectUri
 * @returns {string|null} code_verifier if valid and not expired, otherwise null
 */
function getStoredCodeVerifier(appId, clientId, redirectUri) {
  if (!isBrowser()) return null;
  try {
    var raw = window.sessionStorage.getItem(getStorageKey(appId, clientId, redirectUri));
    if (!raw) return null;
    var _JSON$parse = JSON.parse(raw),
      codeVerifier = _JSON$parse.codeVerifier,
      expiresAt = _JSON$parse.expiresAt;
    if (!codeVerifier || !expiresAt || Date.now() > expiresAt) return null;
    return codeVerifier;
  } catch (_unused) {
    return null;
  }
}

/**
 * @param {string} appId
 * @param {string} clientId
 * @param {string} redirectUri
 * @param {string} codeVerifier
 */
function storeCodeVerifier(appId, clientId, redirectUri, codeVerifier) {
  if (!isBrowser()) return;
  try {
    var key = getStorageKey(appId, clientId, redirectUri);
    var value = JSON.stringify({
      codeVerifier: codeVerifier,
      expiresAt: Date.now() + PKCE_STORAGE_EXPIRY_MS
    });
    window.sessionStorage.setItem(key, value);
  } catch (_unused2) {
    // Ignore storage errors (e.g. private mode); fall back to memory-only
  }
}

/**
 * @param {string} appId
 * @param {string} clientId
 * @param {string} redirectUri
 */
function clearStoredCodeVerifier(appId, clientId, redirectUri) {
  if (!isBrowser()) return;
  try {
    window.sessionStorage.removeItem(getStorageKey(appId, clientId, redirectUri));
  } catch (_unused3) {
    // Ignore
  }
}
