// @contentstack/management/dist/es5/stack/auditlog/index.js
"use strict";

var _interopRequireDefault3 = require("@babel/runtime/helpers/interopRequireDefault");
var _interopRequireDefault2 = _interopRequireDefault3(require("@babel/runtime/helpers/interopRequireDefault"));
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _defineProperty2 = require("@babel/runtime/helpers/defineProperty");
var _defineProperty3 = (0, _interopRequireDefault2["default"])(_defineProperty2);
var _asyncToGenerator2 = require("@babel/runtime/helpers/asyncToGenerator");
var _asyncToGenerator3 = (0, _interopRequireDefault2["default"])(_asyncToGenerator2);
exports.AuditLog = AuditLog;
exports.LogCollection = LogCollection;
var _regenerator = require("@babel/runtime/regenerator");
var _regenerator2 = (0, _interopRequireDefault2["default"])(_regenerator);
var _cloneDeep = require("lodash/cloneDeep");
var _cloneDeep2 = (0, _interopRequireDefault2["default"])(_cloneDeep);
var _contentstackError = require("../../core/contentstackError");
var _contentstackError2 = (0, _interopRequireDefault2["default"])(_contentstackError);
var _entity = require("../../entity");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty3["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
/**
 *
 * @namespace AuditLog
 */
function AuditLog(http) {
  var data = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  this.stackHeaders = data.stackHeaders;
  this.urlPath = "/audit-logs";
  if (data.logs) {
    Object.assign(this, (0, _cloneDeep2["default"])(data.logs));
    this.urlPath = "/audit-logs/".concat(this.uid);

    /**
     * @description The fetch AuditLog call fetches AuditLog details.
     * @memberof AuditLog
     * @func fetch
     * @returns {Promise<Object>} Promise for AuditLog data
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).auditLog('audit_log_item_uid').fetch()
     * .then((log) => console.log(log))
     *
     */
    this.fetch = /*#__PURE__*/(0, _asyncToGenerator3["default"])(/*#__PURE__*/_regenerator2["default"].mark(function _callee() {
      var param,
        headers,
        response,
        _args = arguments,
        _t;
      return _regenerator2["default"].wrap(function (_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            param = _args.length > 0 && _args[0] !== undefined ? _args[0] : {};
            _context.prev = 1;
            headers = {
              headers: _objectSpread({}, (0, _cloneDeep2["default"])(this.stackHeaders)),
              params: _objectSpread({}, (0, _cloneDeep2["default"])(param))
            } || {};
            _context.next = 2;
            return http.get(this.urlPath, headers);
          case 2:
            response = _context.sent;
            if (!response.data) {
              _context.next = 3;
              break;
            }
            return _context.abrupt("return", response.data);
          case 3:
            throw (0, _contentstackError2["default"])(response);
          case 4:
            _context.next = 6;
            break;
          case 5:
            _context.prev = 5;
            _t = _context["catch"](1);
            throw (0, _contentstackError2["default"])(_t);
          case 6:
          case "end":
            return _context.stop();
        }
      }, _callee, this, [[1, 5]]);
    }));
  } else {
    /**
     * @description The Get all AuditLog request retrieves the details of all the AuditLog entries of a stack.
     * @memberof AuditLog
     * @func fetchAll
     * @param {Number} limit The limit parameter will return a specific number of AuditLog entries in the output.
     * @param {Number} skip The skip parameter will skip a specific number of AuditLog entries in the output.
     * @param {Boolean} include_count - To retrieve the count of AuditLog entries.
     * @returns {Promise<ContentstackCollection>} Promise for ContentstackCollection instance.
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).auditLog().fetchAll()
     * .then((logs) => console.log(logs))
     *
     */
    this.fetchAll = (0, _entity.fetchAll)(http, LogCollection);
  }
  return this;
}
function LogCollection(http, data) {
  var obj = (0, _cloneDeep2["default"])(data.logs) || [];
  var logCollection = obj.map(function (userdata) {
    return new AuditLog(http, {
      logs: userdata,
      stackHeaders: data.stackHeaders
    });
  });
  return logCollection;
}
