// @contentstack/management/dist/es-modules/stack/variants/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _regeneratorRuntime from "@babel/runtime/regenerator";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import cloneDeep from 'lodash/cloneDeep';
import { deleteEntity, query } from '../../entity';
import error from '../../core/contentstackError';
/**
 * Variants allow you to group a collection of content within a stack. Using variants you can group content types that need to work together. Read more about <a href='https://www.contentstack.com/docs/developers/create-content-types/manage-variants'>Variants</a>.
 * @namespace Variants
 */
export function Variants(http, data) {
  var _this = this;
  this.stackHeaders = data.stackHeaders;
  this.urlPath = "/variants";
  if (data.variants) {
    Object.assign(this, cloneDeep(data.variants));
    this.urlPath += "/".concat(this.uid);

    /**
     * @description The Delete variant call is used to delete a specific variant.
     * @memberof Variants
     * @func delete
     * @returns {Object} Response Object.
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).variants('variants_uid').delete()
     * .then((response) => console.log(response.notice))
     */
    this["delete"] = deleteEntity(http);

    /**
     * @description The fetch Variant returns information about a particular variant of a stack.
     * @memberof Variants
     * @func fetch
     * @returns {Promise<Variants.Variants>} Promise for Variants instance
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack({ api_key: 'api_key'}).variants('variants_uid').fetch()
     * .then((variants) => console.log(variants))
     *
     */
    this.fetch = /*#__PURE__*/_asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
      var param,
        headers,
        response,
        _args = arguments,
        _t;
      return _regeneratorRuntime.wrap(function (_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            param = _args.length > 0 && _args[0] !== undefined ? _args[0] : {};
            _context.prev = 1;
            headers = {
              headers: _objectSpread({}, cloneDeep(_this.stackHeaders)),
              params: _objectSpread({}, cloneDeep(param))
            } || {};
            _context.next = 2;
            return http.get(_this.urlPath, headers);
          case 2:
            response = _context.sent;
            if (!response.data) {
              _context.next = 3;
              break;
            }
            return _context.abrupt("return", response.data);
          case 3:
            throw error(response);
          case 4:
            _context.next = 6;
            break;
          case 5:
            _context.prev = 5;
            _t = _context["catch"](1);
            error(_t);
          case 6:
          case "end":
            return _context.stop();
        }
      }, _callee, null, [[1, 5]]);
    }));
  } else {
    /**
     * @description The Create a variant call creates a new variant in a particular stack of your Contentstack account.
     * @memberof Variants
     * @func create
     * @returns {Promise<Variants.Variants>} Promise for Variants instance
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     * const variants = {
     *       "uid": "iphone_color_white", // optional
     *       "name": "White",
     *       "personalize_metadata": {
     *           "experience_uid": "exp1",
     *           "experience_short_uid": "expShortUid1",
     *           "project_uid": "project_uid1",
     *           "variant_short_uid": "variantShort_uid1"
     *       }
     * }
     * client.stack().variants().create({ variants })
     * .then((variants) => console.log(variants))
     */
    this.create = /*#__PURE__*/function () {
      var _ref2 = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee2(data) {
        var response, _t2;
        return _regeneratorRuntime.wrap(function (_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              _context2.prev = 0;
              _context2.next = 1;
              return http.post("".concat(_this.urlPath), data, {
                headers: _objectSpread({}, cloneDeep(_this.stackHeaders))
              });
            case 1:
              response = _context2.sent;
              if (!response.data) {
                _context2.next = 2;
                break;
              }
              return _context2.abrupt("return", response.data);
            case 2:
              return _context2.abrupt("return", error(response));
            case 3:
              _context2.next = 5;
              break;
            case 4:
              _context2.prev = 4;
              _t2 = _context2["catch"](0);
              return _context2.abrupt("return", error(_t2));
            case 5:
            case "end":
              return _context2.stop();
          }
        }, _callee2, null, [[0, 4]]);
      }));
      return function (_x) {
        return _ref2.apply(this, arguments);
      };
    }();

    /**
     * @description The Query on Variants will allow you to fetch details of all or specific Variants.
     * @memberof Variants
     * @param {Object} params - URI parameters
     * @prop {Object} params.query - Queries that you can use to fetch filtered results.
     * @func query
     * @returns {Object} Query builder object with find(), count(), and findOne() methods.
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack().variants().query({ query: { name: 'Variants Name' } }).find()
     * .then((variants) => console.log(variants))
     */
    this.query = query({
      http: http,
      wrapperCollection: VariantsCollection
    });

    /**
     * @description The fetchByUIDs call allows you to fetch details of specific Variants by their UIDs.
     * @memberof Variants
     * @param {Array<String>} variantUids - Array of variant UIDs to fetch
     * @func fetchByUIDs
     * @returns {Promise<Object>} Promise for variants data
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     *
     * client.stack().variants().fetchByUIDs(['uid1','uid2'])
     * .then((variants) => console.log(variants))
     */
    this.fetchByUIDs = /*#__PURE__*/function () {
      var _ref3 = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee3(variantUids) {
        var response, _t3;
        return _regeneratorRuntime.wrap(function (_context3) {
          while (1) switch (_context3.prev = _context3.next) {
            case 0:
              _context3.prev = 0;
              _context3.next = 1;
              return http.get(_this.urlPath, {
                params: {
                  uids: variantUids
                },
                headers: _objectSpread({}, cloneDeep(_this.stackHeaders))
              });
            case 1:
              response = _context3.sent;
              if (!response.data) {
                _context3.next = 2;
                break;
              }
              return _context3.abrupt("return", response.data);
            case 2:
              throw error(response);
            case 3:
              _context3.next = 5;
              break;
            case 4:
              _context3.prev = 4;
              _t3 = _context3["catch"](0);
              throw error(_t3);
            case 5:
            case "end":
              return _context3.stop();
          }
        }, _callee3, null, [[0, 4]]);
      }));
      return function (_x2) {
        return _ref3.apply(this, arguments);
      };
    }();
  }
}
export function VariantsCollection(http, data) {
  var obj = cloneDeep(data.variants) || [];
  var VariantsCollection = obj.map(function (userdata) {
    return new Variants(http, {
      variants: userdata,
      stackHeaders: data.stackHeaders
    });
  });
  return VariantsCollection;
}
