// react-select/dist/react-select.browser.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var base_dist_reactSelect = require('./Select-acd252b1.browser.cjs.js');
var stateManager = require('./stateManager-b3381797.browser.cjs.js');
var _inheritsLoose = require('@babel/runtime/helpers/inheritsLoose');
var React = require('react');
var core = require('@emotion/core');
var createCache = require('@emotion/cache');
var memoizeOne = require('memoize-one');
var index$1 = require('./index-dacc4246.browser.cjs.js');
require('@babel/runtime/helpers/objectWithoutPropertiesLoose');
require('@babel/runtime/helpers/readOnlyError');
require('@babel/runtime/helpers/extends');
require('@emotion/css');
require('react-dom');
require('@babel/runtime/helpers/taggedTemplateLiteralLoose');
require('react-input-autosize');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _inheritsLoose__default = /*#__PURE__*/_interopDefault(_inheritsLoose);
var React__default = /*#__PURE__*/_interopDefault(React);
var createCache__default = /*#__PURE__*/_interopDefault(createCache);
var memoizeOne__default = /*#__PURE__*/_interopDefault(memoizeOne);

var NonceProvider = /*#__PURE__*/function (_Component) {
  function NonceProvider(props) {
    var _this;
    _this = _Component.call(this, props) || this;
    _this.createEmotionCache = function (nonce) {
      return createCache__default["default"]({
        nonce: nonce
      });
    };
    _this.createEmotionCache = memoizeOne__default["default"](_this.createEmotionCache);
    return _this;
  }
  _inheritsLoose__default["default"](NonceProvider, _Component);
  var _proto = NonceProvider.prototype;
  _proto.render = function render() {
    var emotionCache = this.createEmotionCache(this.props.nonce);
    return /*#__PURE__*/React__default["default"].createElement(core.CacheProvider, {
      value: emotionCache
    }, this.props.children);
  };
  return NonceProvider;
}(React.Component);

var index = stateManager.manageState(base_dist_reactSelect.Select);

exports.createFilter = base_dist_reactSelect.createFilter;
exports.defaultTheme = base_dist_reactSelect.defaultTheme;
exports.mergeStyles = base_dist_reactSelect.mergeStyles;
exports.components = index$1.components;
exports.NonceProvider = NonceProvider;
exports["default"] = index;
