// react-select/dist/react-select.cjs.prod.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var base_dist_reactSelect = require("./Select-6837a58f.cjs.prod.js"), stateManager = require("./stateManager-b267c635.cjs.prod.js"), _inheritsLoose = require("@babel/runtime/helpers/inheritsLoose"), React = require("react"), core = require("@emotion/core"), createCache = require("@emotion/cache"), memoizeOne = require("memoize-one"), index$1 = require("./index-5bad2954.cjs.prod.js");

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    default: e
  };
}

require("@babel/runtime/helpers/objectWithoutPropertiesLoose"), require("@babel/runtime/helpers/readOnlyError"), 
require("@babel/runtime/helpers/extends"), require("@emotion/css"), require("react-dom"), 
require("@babel/runtime/helpers/taggedTemplateLiteralLoose"), require("react-input-autosize");

var _inheritsLoose__default = _interopDefault(_inheritsLoose), React__default = _interopDefault(React), createCache__default = _interopDefault(createCache), memoizeOne__default = _interopDefault(memoizeOne), NonceProvider = function(_Component) {
  function NonceProvider(props) {
    var _this;
    return (_this = _Component.call(this, props) || this).createEmotionCache = function(nonce) {
      return createCache__default.default({
        nonce: nonce
      });
    }, _this.createEmotionCache = memoizeOne__default.default(_this.createEmotionCache), 
    _this;
  }
  return _inheritsLoose__default.default(NonceProvider, _Component), NonceProvider.prototype.render = function() {
    var emotionCache = this.createEmotionCache(this.props.nonce);
    return React__default.default.createElement(core.CacheProvider, {
      value: emotionCache
    }, this.props.children);
  }, NonceProvider;
}(React.Component), index = stateManager.manageState(base_dist_reactSelect.Select);

exports.createFilter = base_dist_reactSelect.createFilter, exports.defaultTheme = base_dist_reactSelect.defaultTheme, 
exports.mergeStyles = base_dist_reactSelect.mergeStyles, exports.components = index$1.components, 
exports.NonceProvider = NonceProvider, exports.default = index;
