// react-select/dist/react-select.esm.js
import { S as Select } from './Select-09da4243.esm.js';
export { c as createFilter, d as defaultTheme, m as mergeStyles } from './Select-09da4243.esm.js';
import { m as manageState } from './stateManager-64b6fcbb.esm.js';
import _inheritsLoose from '@babel/runtime/helpers/esm/inheritsLoose';
import React, { Component } from 'react';
import { CacheProvider } from '@emotion/core';
import createCache from '@emotion/cache';
import memoizeOne from 'memoize-one';
export { c as components } from './index-b69612cb.esm.js';
import '@babel/runtime/helpers/objectWithoutPropertiesLoose';
import '@babel/runtime/helpers/readOnlyError';
import '@babel/runtime/helpers/extends';
import '@emotion/css';
import 'react-dom';
import '@babel/runtime/helpers/taggedTemplateLiteralLoose';
import 'react-input-autosize';

var NonceProvider = /*#__PURE__*/function (_Component) {
  function NonceProvider(props) {
    var _this;
    _this = _Component.call(this, props) || this;
    _this.createEmotionCache = function (nonce) {
      return createCache({
        nonce: nonce
      });
    };
    _this.createEmotionCache = memoizeOne(_this.createEmotionCache);
    return _this;
  }
  _inheritsLoose(NonceProvider, _Component);
  var _proto = NonceProvider.prototype;
  _proto.render = function render() {
    var emotionCache = this.createEmotionCache(this.props.nonce);
    return /*#__PURE__*/React.createElement(CacheProvider, {
      value: emotionCache
    }, this.props.children);
  };
  return NonceProvider;
}(Component);

var index = manageState(Select);

export { NonceProvider, index as default };
