// react-select/dist/stateManager-b267c635.cjs.prod.js
"use strict";

var _extends = require("@babel/runtime/helpers/extends"), _objectWithoutPropertiesLoose = require("@babel/runtime/helpers/objectWithoutPropertiesLoose"), _inheritsLoose = require("@babel/runtime/helpers/inheritsLoose"), React = require("react");

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    default: e
  };
}

var _extends__default = _interopDefault(_extends), _objectWithoutPropertiesLoose__default = _interopDefault(_objectWithoutPropertiesLoose), _inheritsLoose__default = _interopDefault(_inheritsLoose), React__default = _interopDefault(React), _excluded = [ "defaultInputValue", "defaultMenuIsOpen", "defaultValue" ], defaultProps = {
  defaultInputValue: "",
  defaultMenuIsOpen: !1,
  defaultValue: null
}, manageState = function(SelectComponent) {
  var _StateManager;
  return _StateManager = function(_Component) {
    function StateManager() {
      for (var _this, _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) args[_key] = arguments[_key];
      return (_this = _Component.call.apply(_Component, [ this ].concat(args)) || this).select = void 0, 
      _this.state = {
        inputValue: void 0 !== _this.props.inputValue ? _this.props.inputValue : _this.props.defaultInputValue,
        menuIsOpen: void 0 !== _this.props.menuIsOpen ? _this.props.menuIsOpen : _this.props.defaultMenuIsOpen,
        value: void 0 !== _this.props.value ? _this.props.value : _this.props.defaultValue
      }, _this.onChange = function(value, actionMeta) {
        _this.callProp("onChange", value, actionMeta), _this.setState({
          value: value
        });
      }, _this.onInputChange = function(value, actionMeta) {
        var newValue = _this.callProp("onInputChange", value, actionMeta);
        _this.setState({
          inputValue: void 0 !== newValue ? newValue : value
        });
      }, _this.onMenuOpen = function() {
        _this.callProp("onMenuOpen"), _this.setState({
          menuIsOpen: !0
        });
      }, _this.onMenuClose = function() {
        _this.callProp("onMenuClose"), _this.setState({
          menuIsOpen: !1
        });
      }, _this;
    }
    _inheritsLoose__default.default(StateManager, _Component);
    var _proto = StateManager.prototype;
    return _proto.focus = function() {
      this.select.focus();
    }, _proto.blur = function() {
      this.select.blur();
    }, _proto.getProp = function(key) {
      return void 0 !== this.props[key] ? this.props[key] : this.state[key];
    }, _proto.callProp = function(name) {
      if ("function" == typeof this.props[name]) {
        for (var _this$props, _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) args[_key2 - 1] = arguments[_key2];
        return (_this$props = this.props)[name].apply(_this$props, args);
      }
    }, _proto.render = function() {
      var _this2 = this, _this$props2 = this.props;
      _this$props2.defaultInputValue, _this$props2.defaultMenuIsOpen, _this$props2.defaultValue;
      var props = _objectWithoutPropertiesLoose__default.default(_this$props2, _excluded);
      return React__default.default.createElement(SelectComponent, _extends__default.default({}, props, {
        ref: function(_ref) {
          _this2.select = _ref;
        },
        inputValue: this.getProp("inputValue"),
        menuIsOpen: this.getProp("menuIsOpen"),
        onChange: this.onChange,
        onInputChange: this.onInputChange,
        onMenuClose: this.onMenuClose,
        onMenuOpen: this.onMenuOpen,
        value: this.getProp("value")
      }));
    }, StateManager;
  }(React.Component), _StateManager.defaultProps = defaultProps, _StateManager;
}, manageState$1 = manageState;

exports.manageState = manageState$1;
