// @contentstack/management/dist/es5/contentstackClient.js
"use strict";

var _interopRequireDefault3 = require("@babel/runtime/helpers/interopRequireDefault");
var _interopRequireDefault2 = _interopRequireDefault3(require("@babel/runtime/helpers/interopRequireDefault"));
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _defineProperty2 = require("@babel/runtime/helpers/defineProperty");
var _defineProperty3 = (0, _interopRequireDefault2["default"])(_defineProperty2);
var _objectWithoutProperties2 = require("@babel/runtime/helpers/objectWithoutProperties");
var _objectWithoutProperties3 = (0, _interopRequireDefault2["default"])(_objectWithoutProperties2);
var _excluded = ["mfaSecret"];
/**
 * @namespace ContentstackClient
 */
exports["default"] = contentstackClient;
var _index = require("./stack/index.js");
var _index2 = require("./organization/index");
var _cloneDeep = require("lodash/cloneDeep");
var _cloneDeep2 = (0, _interopRequireDefault2["default"])(_cloneDeep);
var _index3 = require("./user/index");
var _contentstackError = require("./core/contentstackError");
var _contentstackError2 = (0, _interopRequireDefault2["default"])(_contentstackError);
var _oauthHandler = require("./core/oauthHandler");
var _oauthHandler2 = (0, _interopRequireDefault2["default"])(_oauthHandler);
var _otplib = require("otplib");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty3["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function contentstackClient(_ref) {
  var http = _ref.http;
  /**
   * @description The login call is used to sign in to your Contentstack account and obtain the authtoken.
   * @memberof ContentstackClient
   * @func login
   * @param {Object} parameters - login parameters
   * @prop {string} parameters.email - email id for user to login
   * @prop {string} parameters.password - password for user to login
   * @prop {string} parameters.tfa_token - tfa token for user to login (2FA token)
   * @prop {string} parameters.mfaSecret - TOTP secret key for generating 2FA token
   * @param {Object=} params - Optional request parameters
   * @returns {Promise<Object>} Promise for login response
   * @example
   * import * as contentstack from '@contentstack/management'
   * const client = contentstack.client()
   *
   * client.login({ email: <emailid>, password: <password> })
   * .then(() => console.log('Login successful.'))
   *
   * @example
   * client.login({ email: <emailid>, password: <password>, tfa_token: <tfa_token> })
   * .then(() => console.log('Login successful.'))
   *
   * @example
   * client.login({ email: <emailid>, password: <password>, mfaSecret: <mfa_secret> })
   * .then(() => console.log('Login successful.'))
   */
  function login() {
    var requestBody = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    var params = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    http.defaults.versioningStrategy = 'path';
    var _requestBody = requestBody,
      mfaSecret = _requestBody.mfaSecret,
      credentials = (0, _objectWithoutProperties3["default"])(_requestBody, _excluded);
    requestBody = credentials;
    if (!requestBody.tfa_token && mfaSecret) {
      requestBody.tfa_token = _otplib.authenticator.generate(mfaSecret);
    }
    return http.post('/user-session', {
      user: requestBody
    }, {
      params: params
    }).then(function (response) {
      if (response.data.user != null && response.data.user.authtoken != null) {
        http.defaults.headers.common.authtoken = response.data.user.authtoken;
        response.data.user = new _index3.User(http, response.data);
      }
      return response.data;
    }, _contentstackError2["default"]);
  }

  /**
   * @description The Get user call returns comprehensive information of an existing user account.
   * The information returned includes details of the stacks owned by and shared with the specified user account.
   * @memberof ContentstackClient
   * @func getUser
   * @param {Object=} params - Optional request parameters
   * @returns {Promise<Object>} Promise for User instance
   * @example
   * import * as contentstack from '@contentstack/management'
   * const client = contentstack.client()
   *
   * client.getUser()
   * .then((user) => console.log(user))
   *
   */
  function getUser() {
    var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    http.defaults.versioningStrategy = 'path';
    return http.get('/user', {
      params: params
    }).then(function (response) {
      return new _index3.User(http, response.data);
    }, _contentstackError2["default"]);
  }
  /**
   * @description Get Stack instance. A stack is a space that stores the content of a project.
   * @memberof ContentstackClient
   * @func stack
   * @param {Object=} params - Stack initialization parameters
   * @prop {String} params.api_key - Stack API Key
   * @prop {String} params.management_token - Management token for Stack.
   * @prop {String} params.branch_uid - Branch UID or alias to access specific branch. Default is master.
   * @returns {Stack} Instance of Stack.
   *
   * @example
   * import * as contentstack from '@contentstack/management'
   * const client = contentstack.client()
   * const stack = {name: 'My New Stack'}
   * client.stack().create({ stack }, { organization_uid: 'org_uid' })
   * .then((stack) => console.log(stack))
   *
   * @example
   * import * as contentstack from '@contentstack/management'
   * const client = contentstack.client()
   *
   * client.stack({ api_key: 'api_key'}).fetch()
   * .then((stack) => console.log(stack))
   *
   * @example
   * import * as contentstack from '@contentstack/management'
   * const client = contentstack.client()
   *
   * client.stack({ api_key: 'api_key', management_token: 'management_token' }).contentType('content_type_uid').fetch()
   * .then((stack) => console.log(stack))
   *
   * @example
   * import * as contentstack from '@contentstack/management'
   * const client = contentstack.client()
   *
   * client.stack({ api_key: 'api_key', management_token: 'management_token', branch_uid: 'branch_uid' }).contentType('content_type_uid').fetch()
   * .then((stack) => console.log(stack))
   */
  function stack() {
    var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    http.defaults.versioningStrategy = 'path';
    var stack = _objectSpread({}, (0, _cloneDeep2["default"])(params));
    return new _index.Stack(http, {
      stack: stack
    });
  }

  /**
   * @description Organization is the top-level entity in the hierarchy of Contentstack, consisting of stacks and stack resources, and users.
   * @memberof ContentstackClient
   * @func organization
   * @param {String=} uid - Organization UID. If not provided, returns organization instance for querying all organizations.
   * @returns {Organization} Instance of Organization.
   *
   * @example
   * import * as contentstack from '@contentstack/management'
   * const client = contentstack.client()
   *
   * client.organization().findAll()
   * .then((organization) => console.log(organization))
   *
   * @example
   * import * as contentstack from '@contentstack/management'
   * const client = contentstack.client()
   *
   * client.organization('org_uid').fetch()
   * .then((organization) => console.log(organization))
   *
   */
  function organization() {
    var uid = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
    http.defaults.versioningStrategy = 'path';
    return new _index2.Organization(http, uid !== null ? {
      organization: {
        uid: uid
      }
    } : null);
  }

  /**
   * @description The Log out of your account call is used to sign out the user of Contentstack account.
   * @memberof ContentstackClient
   * @param {String} authtoken - Authtoken to logout from.
   * @func logout
   * @returns {Promise<Object>} Promise for response object.
   *
   * @example
   * import * as contentstack from '@contentstack/management'
   * const client = contentstack.client()
   * client.logout()
   * .then((response) => console.log(response))
   *
   * @example
   * import * as contentstack from '@contentstack/management'
   * const client = contentstack.client()
   * client.logout('AUTHTOKEN')
   * .then((response) => console.log(response))
   *   */
  function logout(authtoken) {
    http.defaults.versioningStrategy = 'path';
    if (authtoken !== undefined) {
      return http["delete"]('/user-session', {
        headers: {
          authtoken: authtoken
        }
      }).then(function (response) {
        return response.data;
      }, _contentstackError2["default"]);
    }
    return http["delete"]('/user-session').then(function (response) {
      if (http.defaults.headers.common) {
        delete http.defaults.headers.common.authtoken;
      }
      delete http.defaults.headers.authtoken;
      delete http.httpClientParams.authtoken;
      delete http.httpClientParams.headers.authtoken;
      return response.data;
    }, _contentstackError2["default"]);
  }

  /**
   * @description The oauth call is used to sign in to your Contentstack account and obtain the accesstoken.
   * @memberof ContentstackClient
   * @func oauth
   * @param {Object} parameters - oauth parameters
   * @prop {string=} parameters.appId - appId of the application (default: '6400aa06db64de001a31c8a9')
   * @prop {string=} parameters.clientId - clientId of the application (default: 'Ie0FEfTzlfAHL4xM')
   * @prop {string=} parameters.redirectUri - redirect URI for OAuth callback (default: 'http://localhost:8184')
   * @prop {string=} parameters.responseType - responseType (default: 'code')
   * @prop {string=} parameters.scope - scope
   * @prop {string=} parameters.clientSecret - clientSecret of the application
   * @returns {OAuthHandler} Instance of OAuthHandler.
   * @example
   * import * as contentstack from '@contentstack/management'
   * const client = contentstack.client()
   *
   * client.oauth({ appId: <appId>, clientId: <clientId>, redirectUri: <redirectUri>, clientSecret: <clientSecret>, responseType: <responseType>, scope: <scope> })
   * .then(() => console.log('Login successful.'))
   *
   */
  function oauth() {
    var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    http.defaults.versioningStrategy = 'path';
    var appId = params.appId || '6400aa06db64de001a31c8a9';
    var clientId = params.clientId || 'Ie0FEfTzlfAHL4xM';
    var redirectUri = params.redirectUri || 'http://localhost:8184';
    var responseType = params.responseType || 'code';
    var scope = params.scope;
    var clientSecret = params.clientSecret;
    return new _oauthHandler2["default"](http, appId, clientId, redirectUri, clientSecret, responseType, scope);
  }
  return {
    login: login,
    logout: logout,
    getUser: getUser,
    stack: stack,
    organization: organization,
    axiosInstance: http,
    oauth: oauth
  };
}
module.exports = exports.default;
