// @contentstack/management/dist/es5/contentstackCollection.js
"use strict";

var _interopRequireDefault3 = require("@babel/runtime/helpers/interopRequireDefault");
var _interopRequireDefault2 = _interopRequireDefault3(require("@babel/runtime/helpers/interopRequireDefault"));
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = undefined;
var _defineProperty2 = require("@babel/runtime/helpers/defineProperty");
var _defineProperty3 = (0, _interopRequireDefault2["default"])(_defineProperty2);
var _createClass2 = require("@babel/runtime/helpers/createClass");
var _createClass3 = (0, _interopRequireDefault2["default"])(_createClass2);
var _classCallCheck2 = require("@babel/runtime/helpers/classCallCheck");
var _classCallCheck3 = (0, _interopRequireDefault2["default"])(_classCallCheck2);
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty3["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
/**
 * @namespace ContentstackCollection
 */
var ContentstackCollection = /*#__PURE__*/(0, _createClass3["default"])(
/**
 * Creates a ContentstackCollection instance.
 * @param {Object} response - HTTP response object.
 * @param {Object} http - HTTP client instance.
 * @param {Object=} stackHeaders - Stack headers to include in data.
 * @param {Function=} wrapperCollection - Collection wrapper function to transform items.
 */
function ContentstackCollection(response, http) {
  var _http$httpClientParam;
  var stackHeaders = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
  var wrapperCollection = arguments.length > 3 ? arguments[3] : undefined;
  (0, _classCallCheck3["default"])(this, ContentstackCollection);
  var data = response.data || {};
  if (stackHeaders) {
    data.stackHeaders = stackHeaders;
  }
  if ((http === null || http === void 0 || (_http$httpClientParam = http.httpClientParams) === null || _http$httpClientParam === void 0 || (_http$httpClientParam = _http$httpClientParam.headers) === null || _http$httpClientParam === void 0 ? void 0 : _http$httpClientParam.includeResHeaders) === true) {
    data.stackHeaders = _objectSpread(_objectSpread({}, data.stackHeaders), {}, {
      responseHeaders: response.headers
    });
  }
  if (wrapperCollection) {
    this.items = wrapperCollection(http, data);
  }
  if (data.schema !== undefined) {
    this.schema = data.schema;
  }
  if (data.content_type !== undefined) {
    this.content_type = data.content_type;
  }
  if (data.count !== undefined) {
    this.count = data.count;
  }
  if (data.notice !== undefined) {
    this.notice = data.notice;
  }
});
exports["default"] = ContentstackCollection;
module.exports = exports.default;
