// @contentstack/management/dist/es-modules/stack/contentType/entry/variants/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _regeneratorRuntime from "@babel/runtime/regenerator";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import cloneDeep from 'lodash/cloneDeep';
import { deleteEntity, fetch, query } from '../../../../entity';
import error from '../../../../core/contentstackError';
/**
 * Variants allow you to create variant versions of entries within a content type. Read more about <a href='https://www.contentstack.com/docs/developers/create-content-types/manage-variants'>Variants</a>.
 * @namespace Variants
 */
export function Variants(http, data) {
  var _this = this;
  Object.assign(this, cloneDeep(data));
  this.urlPath = "/content_types/".concat(this.content_type_uid, "/entries/").concat(this.entry_uid, "/variants");
  if (data && data.variants_uid) {
    this.urlPath += "/".concat(this.variants_uid);
    /**
         * @description The Update a variant call updates an existing variant for the selected content type.
         * @memberof Variants
         * @func update
         * @param {Object} data - The variant data to update.
         * @param {Object} [params={}] - Optional query parameters.
         * @returns {Promise<Object>} Response Object.
         * @example
         * import * as contentstack from '@contentstack/management'
         * const client = contentstack.client()
         * const data = {
         *      "entry": {
         *          "title": "example",
         *          "url": "/example",
         *          "_variant": {
         *          "_change_set": [
         *              "title",
         *              "url"
         *              ]
         *          }
         *          }
         *      }
         * client.stack({ api_key: 'api_key'}).contentType('content_type_uid').entry('entry_uid').variants('uid').update(data)
         * .then((variants) => console.log(variants))
         */
    this.update = /*#__PURE__*/function () {
      var _ref = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee(data) {
        var params,
          response,
          _args = arguments,
          _t;
        return _regeneratorRuntime.wrap(function (_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              params = _args.length > 1 && _args[1] !== undefined ? _args[1] : {};
              _context.prev = 1;
              _context.next = 2;
              return http.put(_this.urlPath, data, {
                headers: _objectSpread({}, cloneDeep(_this.stackHeaders)),
                params: _objectSpread({}, cloneDeep(params))
              });
            case 2:
              response = _context.sent;
              if (!response.data) {
                _context.next = 3;
                break;
              }
              return _context.abrupt("return", response.data);
            case 3:
              return _context.abrupt("return", error(response));
            case 4:
              _context.next = 6;
              break;
            case 5:
              _context.prev = 5;
              _t = _context["catch"](1);
              return _context.abrupt("return", error(_t));
            case 6:
            case "end":
              return _context.stop();
          }
        }, _callee, null, [[1, 5]]);
      }));
      return function (_x) {
        return _ref.apply(this, arguments);
      };
    }();

    /**
         * @description The Delete a variant call is used to delete a specific variant from a content type.
         * @memberof Variants
         * @func delete
         * @returns {Promise<Object>} Response Object.
         * @example
         * import * as contentstack from '@contentstack/management'
         * const client = contentstack.client()
         *
         * client.stack({ api_key: 'api_key'}).contentType('content_type_uid').entry('entry_uid').variants('uid').delete()
         * .then((response) => console.log(response.notice))
         */
    this["delete"] = deleteEntity(http);

    /**
         * @description The fetch Variants call fetches Variants details.
         * @memberof Variants
         * @func fetch
         * @returns {Promise<Variants>} Promise for Variants instance
         * @example
         * import * as contentstack from '@contentstack/management'
         * const client = contentstack.client()
         *
         * client.stack({ api_key: 'api_key'}).contentType('content_type_uid').entry('entry_uid').variants('uid').fetch()
         * .then((variants) => console.log(variants))
         *
         */
    this.fetch = fetch(http, 'variants');

    /**
         * @description The versions call fetches variant version details.
         * @memberof Variants
         * @func versions
         * @returns {Promise<Object>} Promise for variant versions data
         * @example
         * import * as contentstack from '@contentstack/management'
         * const client = contentstack.client()
         *
         * client.stack({ api_key: 'api_key'}).contentType('content_type_uid').entry('entry_uid').variants('uid').versions()
         * .then((variants) => console.log(variants))
         *
         */
    this.versions = /*#__PURE__*/_asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee2() {
      var response, _t2;
      return _regeneratorRuntime.wrap(function (_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            _context2.prev = 0;
            _context2.next = 1;
            return http.get("".concat(_this.urlPath, "/versions"), {
              headers: _objectSpread({}, cloneDeep(_this.stackHeaders))
            });
          case 1:
            response = _context2.sent;
            if (!response.data) {
              _context2.next = 2;
              break;
            }
            return _context2.abrupt("return", response.data);
          case 2:
            return _context2.abrupt("return", error(response));
          case 3:
            _context2.next = 5;
            break;
          case 4:
            _context2.prev = 4;
            _t2 = _context2["catch"](0);
            return _context2.abrupt("return", error(_t2));
          case 5:
          case "end":
            return _context2.stop();
        }
      }, _callee2, null, [[0, 4]]);
    }));
  } else {
    /**
         * @description The Query on Variants will allow you to fetch details of all or specific Variants.
         * @memberof Variants
         * @func query
         * @param {Object} params - Query parameters
         * @prop {Int} params.locale Enter the code of the language of which the variants need to be included.
         * @prop {Object} params.query Queries that you can use to fetch filtered results.
         * @returns {Object} Query builder object with find(), count(), and findOne() methods.
         *
         * @example
         * import * as contentstack from '@contentstack/management'
         * const client = contentstack.client()
         *
         * client.stack().contentType('content_type_uid').entry('entry_uid').variants().query({ query: { title: 'Variants title' } }).find()
         * .then((variants) => console.log(variants))
         */
    this.query = query({
      http: http,
      wrapperCollection: VariantsCollection
    });
  }
}
export function VariantsCollection(http, data) {
  var obj = cloneDeep(data.entries) || [];
  var variantCollection = obj.map(function (variant) {
    return new Variants(http, {
      content_type_uid: data.content_type_uid,
      entry_uid: variant.uid,
      variants_uid: variant._variant._uid,
      stackHeaders: data.stackHeaders,
      variants: variant
    });
  });
  return variantCollection;
}
