// react-select/creatable/dist/react-select.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _inheritsLoose = require('@babel/runtime/helpers/inheritsLoose');
var _extends = require('@babel/runtime/helpers/extends');
var React = require('react');
var base_dist_reactSelect = require('../../dist/Select-3f8c62fe.cjs.dev.js');
var index = require('../../dist/index-b61114b2.cjs.dev.js');
var stateManager = require('../../dist/stateManager-5134c53b.cjs.dev.js');
require('@babel/runtime/helpers/objectWithoutPropertiesLoose');
require('@babel/runtime/helpers/readOnlyError');
require('memoize-one');
require('@emotion/css');
require('@emotion/core');
require('react-dom');
require('@babel/runtime/helpers/taggedTemplateLiteralLoose');
require('react-input-autosize');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _inheritsLoose__default = /*#__PURE__*/_interopDefault(_inheritsLoose);
var _extends__default = /*#__PURE__*/_interopDefault(_extends);
var React__default = /*#__PURE__*/_interopDefault(React);

var compareOption = function compareOption(inputValue, option) {
  if (inputValue === void 0) {
    inputValue = '';
  }
  var candidate = String(inputValue).toLowerCase();
  var optionValue = String(option.value).toLowerCase();
  var optionLabel = String(option.label).toLowerCase();
  return optionValue === candidate || optionLabel === candidate;
};
var builtins = {
  formatCreateLabel: function formatCreateLabel(inputValue) {
    return "Create \"" + inputValue + "\"";
  },
  isValidNewOption: function isValidNewOption(inputValue, selectValue, selectOptions) {
    return !(!inputValue || selectValue.some(function (option) {
      return compareOption(inputValue, option);
    }) || selectOptions.some(function (option) {
      return compareOption(inputValue, option);
    }));
  },
  getNewOptionData: function getNewOptionData(inputValue, optionLabel) {
    return {
      label: optionLabel,
      value: inputValue,
      __isNew__: true
    };
  }
};
var defaultProps = _extends__default["default"]({
  allowCreateWhileLoading: false,
  createOptionPosition: 'last'
}, builtins);
var makeCreatableSelect = function makeCreatableSelect(SelectComponent) {
  var _Creatable;
  return _Creatable = /*#__PURE__*/function (_Component) {
    function Creatable(props) {
      var _this;
      _this = _Component.call(this, props) || this;
      _this.select = void 0;
      _this.onChange = function (newValue, actionMeta) {
        var _this$props = _this.props,
          getNewOptionData = _this$props.getNewOptionData,
          inputValue = _this$props.inputValue,
          isMulti = _this$props.isMulti,
          onChange = _this$props.onChange,
          onCreateOption = _this$props.onCreateOption,
          value = _this$props.value,
          name = _this$props.name;
        if (actionMeta.action !== 'select-option') {
          return onChange(newValue, actionMeta);
        }
        var newOption = _this.state.newOption;
        var valueArray = Array.isArray(newValue) ? newValue : [newValue];
        if (valueArray[valueArray.length - 1] === newOption) {
          if (onCreateOption) onCreateOption(inputValue);else {
            var newOptionData = getNewOptionData(inputValue, inputValue);
            var newActionMeta = {
              action: 'create-option',
              name: name
            };
            if (isMulti) {
              onChange([].concat(index.cleanValue(value), [newOptionData]), newActionMeta);
            } else {
              onChange(newOptionData, newActionMeta);
            }
          }
          return;
        }
        onChange(newValue, actionMeta);
      };
      var options = props.options || [];
      _this.state = {
        newOption: undefined,
        options: options
      };
      return _this;
    }
    _inheritsLoose__default["default"](Creatable, _Component);
    var _proto = Creatable.prototype;
    _proto.UNSAFE_componentWillReceiveProps = function UNSAFE_componentWillReceiveProps(nextProps) {
      var allowCreateWhileLoading = nextProps.allowCreateWhileLoading,
        createOptionPosition = nextProps.createOptionPosition,
        formatCreateLabel = nextProps.formatCreateLabel,
        getNewOptionData = nextProps.getNewOptionData,
        inputValue = nextProps.inputValue,
        isLoading = nextProps.isLoading,
        isValidNewOption = nextProps.isValidNewOption,
        value = nextProps.value;
      var options = nextProps.options || [];
      var newOption = this.state.newOption;
      if (isValidNewOption(inputValue, index.cleanValue(value), options)) {
        newOption = getNewOptionData(inputValue, formatCreateLabel(inputValue));
      } else {
        newOption = undefined;
      }
      this.setState({
        newOption: newOption,
        options: (allowCreateWhileLoading || !isLoading) && newOption ? createOptionPosition === 'first' ? [newOption].concat(options) : [].concat(options, [newOption]) : options
      });
    };
    _proto.focus = function focus() {
      this.select.focus();
    };
    _proto.blur = function blur() {
      this.select.blur();
    };
    _proto.render = function render() {
      var _this2 = this;
      var options = this.state.options;
      return /*#__PURE__*/React__default["default"].createElement(SelectComponent, _extends__default["default"]({}, this.props, {
        ref: function ref(_ref) {
          _this2.select = _ref;
        },
        options: options,
        onChange: this.onChange
      }));
    };
    return Creatable;
  }(React.Component), _Creatable.defaultProps = defaultProps, _Creatable;
};

// TODO: do this in package entrypoint
var SelectCreatable = makeCreatableSelect(base_dist_reactSelect.Select);
var Creatable = stateManager.manageState(SelectCreatable);

exports["default"] = Creatable;
exports.defaultProps = defaultProps;
exports.makeCreatableSelect = makeCreatableSelect;
