// react-select/creatable/dist/react-select.cjs.prod.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var _inheritsLoose = require("@babel/runtime/helpers/inheritsLoose"), _extends = require("@babel/runtime/helpers/extends"), React = require("react"), base_dist_reactSelect = require("../../dist/Select-6837a58f.cjs.prod.js"), index = require("../../dist/index-5bad2954.cjs.prod.js"), stateManager = require("../../dist/stateManager-b267c635.cjs.prod.js");

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    default: e
  };
}

require("@babel/runtime/helpers/objectWithoutPropertiesLoose"), require("@babel/runtime/helpers/readOnlyError"), 
require("memoize-one"), require("@emotion/css"), require("@emotion/core"), require("react-dom"), 
require("@babel/runtime/helpers/taggedTemplateLiteralLoose"), require("react-input-autosize");

var _inheritsLoose__default = _interopDefault(_inheritsLoose), _extends__default = _interopDefault(_extends), React__default = _interopDefault(React), compareOption = function(inputValue, option) {
  void 0 === inputValue && (inputValue = "");
  var candidate = String(inputValue).toLowerCase(), optionValue = String(option.value).toLowerCase(), optionLabel = String(option.label).toLowerCase();
  return optionValue === candidate || optionLabel === candidate;
}, builtins = {
  formatCreateLabel: function(inputValue) {
    return 'Create "' + inputValue + '"';
  },
  isValidNewOption: function(inputValue, selectValue, selectOptions) {
    return !(!inputValue || selectValue.some(function(option) {
      return compareOption(inputValue, option);
    }) || selectOptions.some(function(option) {
      return compareOption(inputValue, option);
    }));
  },
  getNewOptionData: function(inputValue, optionLabel) {
    return {
      label: optionLabel,
      value: inputValue,
      __isNew__: !0
    };
  }
}, defaultProps = _extends__default.default({
  allowCreateWhileLoading: !1,
  createOptionPosition: "last"
}, builtins), makeCreatableSelect = function(SelectComponent) {
  var _Creatable;
  return _Creatable = function(_Component) {
    function Creatable(props) {
      var _this;
      (_this = _Component.call(this, props) || this).select = void 0, _this.onChange = function(newValue, actionMeta) {
        var _this$props = _this.props, getNewOptionData = _this$props.getNewOptionData, inputValue = _this$props.inputValue, isMulti = _this$props.isMulti, onChange = _this$props.onChange, onCreateOption = _this$props.onCreateOption, value = _this$props.value, name = _this$props.name;
        if ("select-option" !== actionMeta.action) return onChange(newValue, actionMeta);
        var newOption = _this.state.newOption, valueArray = Array.isArray(newValue) ? newValue : [ newValue ];
        if (valueArray[valueArray.length - 1] !== newOption) onChange(newValue, actionMeta); else if (onCreateOption) onCreateOption(inputValue); else {
          var newOptionData = getNewOptionData(inputValue, inputValue), newActionMeta = {
            action: "create-option",
            name: name
          };
          onChange(isMulti ? [].concat(index.cleanValue(value), [ newOptionData ]) : newOptionData, newActionMeta);
        }
      };
      var options = props.options || [];
      return _this.state = {
        newOption: void 0,
        options: options
      }, _this;
    }
    _inheritsLoose__default.default(Creatable, _Component);
    var _proto = Creatable.prototype;
    return _proto.UNSAFE_componentWillReceiveProps = function(nextProps) {
      var allowCreateWhileLoading = nextProps.allowCreateWhileLoading, createOptionPosition = nextProps.createOptionPosition, formatCreateLabel = nextProps.formatCreateLabel, getNewOptionData = nextProps.getNewOptionData, inputValue = nextProps.inputValue, isLoading = nextProps.isLoading, isValidNewOption = nextProps.isValidNewOption, value = nextProps.value, options = nextProps.options || [], newOption = this.state.newOption;
      newOption = isValidNewOption(inputValue, index.cleanValue(value), options) ? getNewOptionData(inputValue, formatCreateLabel(inputValue)) : void 0, 
      this.setState({
        newOption: newOption,
        options: !allowCreateWhileLoading && isLoading || !newOption ? options : "first" === createOptionPosition ? [ newOption ].concat(options) : [].concat(options, [ newOption ])
      });
    }, _proto.focus = function() {
      this.select.focus();
    }, _proto.blur = function() {
      this.select.blur();
    }, _proto.render = function() {
      var _this2 = this, options = this.state.options;
      return React__default.default.createElement(SelectComponent, _extends__default.default({}, this.props, {
        ref: function(_ref) {
          _this2.select = _ref;
        },
        options: options,
        onChange: this.onChange
      }));
    }, Creatable;
  }(React.Component), _Creatable.defaultProps = defaultProps, _Creatable;
}, SelectCreatable = makeCreatableSelect(base_dist_reactSelect.Select), Creatable = stateManager.manageState(SelectCreatable);

exports.default = Creatable, exports.defaultProps = defaultProps, exports.makeCreatableSelect = makeCreatableSelect;
