// react-select/async/dist/react-select.cjs.prod.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var _extends = require("@babel/runtime/helpers/extends"), _objectWithoutPropertiesLoose = require("@babel/runtime/helpers/objectWithoutPropertiesLoose");

require("@babel/runtime/helpers/readOnlyError");

var _inheritsLoose = require("@babel/runtime/helpers/inheritsLoose"), React = require("react"), base_dist_reactSelect = require("../../dist/Select-6837a58f.cjs.prod.js"), index = require("../../dist/index-5bad2954.cjs.prod.js"), stateManager = require("../../dist/stateManager-b267c635.cjs.prod.js");

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    default: e
  };
}

require("memoize-one"), require("@emotion/css"), require("@emotion/core"), require("react-dom"), 
require("@babel/runtime/helpers/taggedTemplateLiteralLoose"), require("react-input-autosize");

var _extends__default = _interopDefault(_extends), _objectWithoutPropertiesLoose__default = _interopDefault(_objectWithoutPropertiesLoose), _inheritsLoose__default = _interopDefault(_inheritsLoose), React__default = _interopDefault(React), _excluded = [ "loadOptions", "isLoading" ], defaultProps = {
  cacheOptions: !1,
  defaultOptions: !1,
  filterOption: null,
  isLoading: !1
}, makeAsyncSelect = function(SelectComponent) {
  var _Async;
  return _Async = function(_Component) {
    function Async(props) {
      var _this;
      return (_this = _Component.call(this) || this).select = void 0, _this.lastRequest = void 0, 
      _this.mounted = !1, _this.optionsCache = {}, _this.handleInputChange = function(newValue, actionMeta) {
        var _this$props = _this.props, cacheOptions = _this$props.cacheOptions, onInputChange = _this$props.onInputChange, inputValue = index.handleInputChange(newValue, actionMeta, onInputChange);
        if (!inputValue) return delete _this.lastRequest, void _this.setState({
          inputValue: "",
          loadedInputValue: "",
          loadedOptions: [],
          isLoading: !1,
          passEmptyOptions: !1
        });
        if (cacheOptions && _this.optionsCache[inputValue]) _this.setState({
          inputValue: inputValue,
          loadedInputValue: inputValue,
          loadedOptions: _this.optionsCache[inputValue],
          isLoading: !1,
          passEmptyOptions: !1
        }); else {
          var request = _this.lastRequest = {};
          _this.setState({
            inputValue: inputValue,
            isLoading: !0,
            passEmptyOptions: !_this.state.loadedInputValue
          }, function() {
            _this.loadOptions(inputValue, function(options) {
              _this.mounted && (options && (_this.optionsCache[inputValue] = options), request === _this.lastRequest && (delete _this.lastRequest, 
              _this.setState({
                isLoading: !1,
                loadedInputValue: inputValue,
                loadedOptions: options || [],
                passEmptyOptions: !1
              })));
            });
          });
        }
        return inputValue;
      }, _this.state = {
        defaultOptions: Array.isArray(props.defaultOptions) ? props.defaultOptions : void 0,
        inputValue: void 0 !== props.inputValue ? props.inputValue : "",
        isLoading: !0 === props.defaultOptions,
        loadedOptions: [],
        passEmptyOptions: !1
      }, _this;
    }
    _inheritsLoose__default.default(Async, _Component);
    var _proto = Async.prototype;
    return _proto.componentDidMount = function() {
      var _this2 = this;
      this.mounted = !0;
      var defaultOptions = this.props.defaultOptions, inputValue = this.state.inputValue;
      !0 === defaultOptions && this.loadOptions(inputValue, function(options) {
        if (_this2.mounted) {
          var isLoading = !!_this2.lastRequest;
          _this2.setState({
            defaultOptions: options || [],
            isLoading: isLoading
          });
        }
      });
    }, _proto.UNSAFE_componentWillReceiveProps = function(nextProps) {
      nextProps.cacheOptions !== this.props.cacheOptions && (this.optionsCache = {}), 
      nextProps.defaultOptions !== this.props.defaultOptions && this.setState({
        defaultOptions: Array.isArray(nextProps.defaultOptions) ? nextProps.defaultOptions : void 0
      });
    }, _proto.componentWillUnmount = function() {
      this.mounted = !1;
    }, _proto.focus = function() {
      this.select.focus();
    }, _proto.blur = function() {
      this.select.blur();
    }, _proto.loadOptions = function(inputValue, callback) {
      var loadOptions = this.props.loadOptions;
      if (!loadOptions) return callback();
      var loader = loadOptions(inputValue, callback);
      loader && "function" == typeof loader.then && loader.then(callback, function() {
        return callback();
      });
    }, _proto.render = function() {
      var _this3 = this, _this$props2 = this.props;
      _this$props2.loadOptions;
      var isLoadingProp = _this$props2.isLoading, props = _objectWithoutPropertiesLoose__default.default(_this$props2, _excluded), _this$state = this.state, defaultOptions = _this$state.defaultOptions, inputValue = _this$state.inputValue, isLoading = _this$state.isLoading, loadedInputValue = _this$state.loadedInputValue, loadedOptions = _this$state.loadedOptions, options = _this$state.passEmptyOptions ? [] : inputValue && loadedInputValue ? loadedOptions : defaultOptions || [];
      return React__default.default.createElement(SelectComponent, _extends__default.default({}, props, {
        ref: function(_ref) {
          _this3.select = _ref;
        },
        options: options,
        isLoading: isLoading || isLoadingProp,
        onInputChange: this.handleInputChange
      }));
    }, Async;
  }(React.Component), _Async.defaultProps = defaultProps, _Async;
}, SelectState = stateManager.manageState(base_dist_reactSelect.Select), Async = makeAsyncSelect(SelectState);

exports.default = Async, exports.defaultProps = defaultProps, exports.makeAsyncSelect = makeAsyncSelect;
