// react-select/async/dist/react-select.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime/helpers/extends');
var _objectWithoutPropertiesLoose = require('@babel/runtime/helpers/objectWithoutPropertiesLoose');
require('@babel/runtime/helpers/readOnlyError');
var _inheritsLoose = require('@babel/runtime/helpers/inheritsLoose');
var React = require('react');
var base_dist_reactSelect = require('../../dist/Select-3f8c62fe.cjs.dev.js');
var index = require('../../dist/index-b61114b2.cjs.dev.js');
var stateManager = require('../../dist/stateManager-5134c53b.cjs.dev.js');
require('memoize-one');
require('@emotion/css');
require('@emotion/core');
require('react-dom');
require('@babel/runtime/helpers/taggedTemplateLiteralLoose');
require('react-input-autosize');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _extends__default = /*#__PURE__*/_interopDefault(_extends);
var _objectWithoutPropertiesLoose__default = /*#__PURE__*/_interopDefault(_objectWithoutPropertiesLoose);
var _inheritsLoose__default = /*#__PURE__*/_interopDefault(_inheritsLoose);
var React__default = /*#__PURE__*/_interopDefault(React);

var _excluded = ["loadOptions", "isLoading"];
var defaultProps = {
  cacheOptions: false,
  defaultOptions: false,
  filterOption: null,
  isLoading: false
};
var makeAsyncSelect = function makeAsyncSelect(SelectComponent) {
  var _Async;
  return _Async = /*#__PURE__*/function (_Component) {
    function Async(props) {
      var _this;
      _this = _Component.call(this) || this;
      _this.select = void 0;
      _this.lastRequest = void 0;
      _this.mounted = false;
      _this.optionsCache = {};
      _this.handleInputChange = function (newValue, actionMeta) {
        var _this$props = _this.props,
          cacheOptions = _this$props.cacheOptions,
          onInputChange = _this$props.onInputChange;
        // TODO
        var inputValue = index.handleInputChange(newValue, actionMeta, onInputChange);
        if (!inputValue) {
          delete _this.lastRequest;
          _this.setState({
            inputValue: '',
            loadedInputValue: '',
            loadedOptions: [],
            isLoading: false,
            passEmptyOptions: false
          });
          return;
        }
        if (cacheOptions && _this.optionsCache[inputValue]) {
          _this.setState({
            inputValue: inputValue,
            loadedInputValue: inputValue,
            loadedOptions: _this.optionsCache[inputValue],
            isLoading: false,
            passEmptyOptions: false
          });
        } else {
          var request = _this.lastRequest = {};
          _this.setState({
            inputValue: inputValue,
            isLoading: true,
            passEmptyOptions: !_this.state.loadedInputValue
          }, function () {
            _this.loadOptions(inputValue, function (options) {
              if (!_this.mounted) return;
              if (options) {
                _this.optionsCache[inputValue] = options;
              }
              if (request !== _this.lastRequest) return;
              delete _this.lastRequest;
              _this.setState({
                isLoading: false,
                loadedInputValue: inputValue,
                loadedOptions: options || [],
                passEmptyOptions: false
              });
            });
          });
        }
        return inputValue;
      };
      _this.state = {
        defaultOptions: Array.isArray(props.defaultOptions) ? props.defaultOptions : undefined,
        inputValue: typeof props.inputValue !== 'undefined' ? props.inputValue : '',
        isLoading: props.defaultOptions === true,
        loadedOptions: [],
        passEmptyOptions: false
      };
      return _this;
    }
    _inheritsLoose__default["default"](Async, _Component);
    var _proto = Async.prototype;
    _proto.componentDidMount = function componentDidMount() {
      var _this2 = this;
      this.mounted = true;
      var defaultOptions = this.props.defaultOptions;
      var inputValue = this.state.inputValue;
      if (defaultOptions === true) {
        this.loadOptions(inputValue, function (options) {
          if (!_this2.mounted) return;
          var isLoading = !!_this2.lastRequest;
          _this2.setState({
            defaultOptions: options || [],
            isLoading: isLoading
          });
        });
      }
    };
    _proto.UNSAFE_componentWillReceiveProps = function UNSAFE_componentWillReceiveProps(nextProps) {
      // if the cacheOptions prop changes, clear the cache
      if (nextProps.cacheOptions !== this.props.cacheOptions) {
        this.optionsCache = {};
      }
      if (nextProps.defaultOptions !== this.props.defaultOptions) {
        this.setState({
          defaultOptions: Array.isArray(nextProps.defaultOptions) ? nextProps.defaultOptions : undefined
        });
      }
    };
    _proto.componentWillUnmount = function componentWillUnmount() {
      this.mounted = false;
    };
    _proto.focus = function focus() {
      this.select.focus();
    };
    _proto.blur = function blur() {
      this.select.blur();
    };
    _proto.loadOptions = function loadOptions(inputValue, callback) {
      var loadOptions = this.props.loadOptions;
      if (!loadOptions) return callback();
      var loader = loadOptions(inputValue, callback);
      if (loader && typeof loader.then === 'function') {
        loader.then(callback, function () {
          return callback();
        });
      }
    };
    _proto.render = function render() {
      var _this3 = this;
      var _this$props2 = this.props;
        _this$props2.loadOptions;
        var isLoadingProp = _this$props2.isLoading,
        props = _objectWithoutPropertiesLoose__default["default"](_this$props2, _excluded);
      var _this$state = this.state,
        defaultOptions = _this$state.defaultOptions,
        inputValue = _this$state.inputValue,
        isLoading = _this$state.isLoading,
        loadedInputValue = _this$state.loadedInputValue,
        loadedOptions = _this$state.loadedOptions,
        passEmptyOptions = _this$state.passEmptyOptions;
      var options = passEmptyOptions ? [] : inputValue && loadedInputValue ? loadedOptions : defaultOptions || [];
      return /*#__PURE__*/React__default["default"].createElement(SelectComponent, _extends__default["default"]({}, props, {
        ref: function ref(_ref) {
          _this3.select = _ref;
        },
        options: options,
        isLoading: isLoading || isLoadingProp,
        onInputChange: this.handleInputChange
      }));
    };
    return Async;
  }(React.Component), _Async.defaultProps = defaultProps, _Async;
};
var SelectState = stateManager.manageState(base_dist_reactSelect.Select);
var Async = makeAsyncSelect(SelectState);

exports["default"] = Async;
exports.defaultProps = defaultProps;
exports.makeAsyncSelect = makeAsyncSelect;
