// react-select/animated/dist/react-select.cjs.prod.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var _extends = require("@babel/runtime/helpers/extends"), _objectWithoutPropertiesLoose = require("@babel/runtime/helpers/objectWithoutPropertiesLoose"), memoizeOne = require("memoize-one"), index$1 = require("../../dist/index-5bad2954.cjs.prod.js"), React = require("react"), _inheritsLoose = require("@babel/runtime/helpers/inheritsLoose"), reactTransitionGroup = require("react-transition-group");

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    default: e
  };
}

require("@emotion/core"), require("@emotion/css"), require("@babel/runtime/helpers/taggedTemplateLiteralLoose"), 
require("react-input-autosize"), require("react-dom");

var _extends__default = _interopDefault(_extends), _objectWithoutPropertiesLoose__default = _interopDefault(_objectWithoutPropertiesLoose), memoizeOne__default = _interopDefault(memoizeOne), React__default = _interopDefault(React), _inheritsLoose__default = _interopDefault(_inheritsLoose), _excluded$3 = [ "in", "onExited", "appear", "enter", "exit" ], AnimatedInput = function(WrappedComponent) {
  return function(_ref) {
    _ref.in, _ref.onExited, _ref.appear, _ref.enter, _ref.exit;
    var props = _objectWithoutPropertiesLoose__default.default(_ref, _excluded$3);
    return React__default.default.createElement(WrappedComponent, props);
  };
}, AnimatedInput$1 = AnimatedInput, _excluded$2 = [ "component", "duration", "in", "onExited" ], Fade = function(_ref) {
  var Tag = _ref.component, _ref$duration = _ref.duration, duration = void 0 === _ref$duration ? 1 : _ref$duration, inProp = _ref.in;
  _ref.onExited;
  var props = _objectWithoutPropertiesLoose__default.default(_ref, _excluded$2), transition = {
    entering: {
      opacity: 0
    },
    entered: {
      opacity: 1,
      transition: "opacity " + duration + "ms"
    },
    exiting: {
      opacity: 0
    },
    exited: {
      opacity: 0
    }
  };
  return React__default.default.createElement(reactTransitionGroup.Transition, {
    mountOnEnter: !0,
    unmountOnExit: !0,
    in: inProp,
    timeout: duration
  }, function(state) {
    var innerProps = {
      style: _extends__default.default({}, transition[state])
    };
    return React__default.default.createElement(Tag, _extends__default.default({
      innerProps: innerProps
    }, props));
  });
}, collapseDuration = 260, Collapse = function(_Component) {
  function Collapse() {
    for (var _this, _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) args[_key] = arguments[_key];
    return (_this = _Component.call.apply(_Component, [ this ].concat(args)) || this).duration = collapseDuration, 
    _this.rafID = void 0, _this.state = {
      width: "auto"
    }, _this.transition = {
      exiting: {
        width: 0,
        transition: "width " + _this.duration + "ms ease-out"
      },
      exited: {
        width: 0
      }
    }, _this.getWidth = function(ref) {
      ref && isNaN(_this.state.width) && (_this.rafID = window.requestAnimationFrame(function() {
        var width = ref.getBoundingClientRect().width;
        _this.setState({
          width: width
        });
      }));
    }, _this.getStyle = function(width) {
      return {
        overflow: "hidden",
        whiteSpace: "nowrap",
        width: width
      };
    }, _this.getTransition = function(state) {
      return _this.transition[state];
    }, _this;
  }
  _inheritsLoose__default.default(Collapse, _Component);
  var _proto = Collapse.prototype;
  return _proto.componentWillUnmount = function() {
    this.rafID && window.cancelAnimationFrame(this.rafID);
  }, _proto.render = function() {
    var _this2 = this, _this$props = this.props, children = _this$props.children, inProp = _this$props.in, width = this.state.width;
    return React__default.default.createElement(reactTransitionGroup.Transition, {
      enter: !1,
      mountOnEnter: !0,
      unmountOnExit: !0,
      in: inProp,
      timeout: this.duration
    }, function(state) {
      var style = _extends__default.default({}, _this2.getStyle(width), _this2.getTransition(state));
      return React__default.default.createElement("div", {
        ref: _this2.getWidth,
        style: style
      }, children);
    });
  }, Collapse;
}(React.Component), _excluded$1 = [ "in", "onExited" ], AnimatedMultiValue = function(WrappedComponent) {
  return function(_ref) {
    var inProp = _ref.in, onExited = _ref.onExited, props = _objectWithoutPropertiesLoose__default.default(_ref, _excluded$1);
    return React__default.default.createElement(Collapse, {
      in: inProp,
      onExited: onExited
    }, React__default.default.createElement(WrappedComponent, _extends__default.default({
      cropWithEllipsis: inProp
    }, props)));
  };
}, AnimatedMultiValue$1 = AnimatedMultiValue, AnimatedPlaceholder = function(WrappedComponent) {
  return function(props) {
    return React__default.default.createElement(Fade, _extends__default.default({
      component: WrappedComponent,
      duration: props.isMulti ? collapseDuration : 1
    }, props));
  };
}, AnimatedPlaceholder$1 = AnimatedPlaceholder, AnimatedSingleValue = function(WrappedComponent) {
  return function(props) {
    return React__default.default.createElement(Fade, _extends__default.default({
      component: WrappedComponent
    }, props));
  };
}, AnimatedSingleValue$1 = AnimatedSingleValue, AnimatedValueContainer = function(WrappedComponent) {
  return function(props) {
    return React__default.default.createElement(reactTransitionGroup.TransitionGroup, _extends__default.default({
      component: WrappedComponent
    }, props));
  };
}, AnimatedValueContainer$1 = AnimatedValueContainer, _excluded = [ "Input", "MultiValue", "Placeholder", "SingleValue", "ValueContainer" ], makeAnimated = function(externalComponents) {
  void 0 === externalComponents && (externalComponents = {});
  var components = index$1.defaultComponents({
    components: externalComponents
  }), Input = components.Input, MultiValue = components.MultiValue, Placeholder = components.Placeholder, SingleValue = components.SingleValue, ValueContainer = components.ValueContainer, rest = _objectWithoutPropertiesLoose__default.default(components, _excluded);
  return _extends__default.default({
    Input: AnimatedInput$1(Input),
    MultiValue: AnimatedMultiValue$1(MultiValue),
    Placeholder: AnimatedPlaceholder$1(Placeholder),
    SingleValue: AnimatedSingleValue$1(SingleValue),
    ValueContainer: AnimatedValueContainer$1(ValueContainer)
  }, rest);
}, AnimatedComponents = makeAnimated(), Input = AnimatedComponents.Input, MultiValue = AnimatedComponents.MultiValue, Placeholder = AnimatedComponents.Placeholder, SingleValue = AnimatedComponents.SingleValue, ValueContainer = AnimatedComponents.ValueContainer, index = memoizeOne__default.default(makeAnimated, index$1.exportedEqual);

exports.Input = Input, exports.MultiValue = MultiValue, exports.Placeholder = Placeholder, 
exports.SingleValue = SingleValue, exports.ValueContainer = ValueContainer, exports.default = index;
